<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Managing Open and Reproducible Computational Projects: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
          Â 
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Managing Open and Reproducible Computational Projects
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Managing Open and Reproducible Computational Projects
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="reference.html"></a></li>
<li><a class="dropdown-item" href="04-FAIRHowTo.html">FAIR How-Tos for data and software</a></li>
<li><a class="dropdown-item" href="08-KanbanTutorial.html">Kanban Boards for Project Management</a></li>
<li><a class="dropdown-item" href="09-GithubCodeReview.html">What Code Review Looks Like in GitHub</a></li>
<li><a class="dropdown-item" href="51-datawrangling.html">Data Wrangling and exploration</a></li>
<li><a class="dropdown-item" href="spreadsheets.html">Details on spreadsheets design</a></li>
<li><a class="dropdown-item" href="16-open.html">Open Science Practices</a></li>
<li><a class="dropdown-item" href="statistics.html">statistics, some details</a></li>
          </ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Managing Open and Reproducible Computational Projects
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="21_Definitions.html">1. Project management, data science</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="22_contentAndMotivation.html">2. Course content and motivation</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="23_Ensuring-reproducibility.html">3. Ensuring reproducibility</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="24_SettingUp.html">4. Managing project start and collaborations</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="25_RDM.html">5. Managing Data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="26_Coding.html">6. Managing code</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="27_publications.html">7. Managing publication</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="28_conclusion.html">8. Conclusion and feedback</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html"></a></li>
<li><a href="04-FAIRHowTo.html">FAIR How-Tos for data and software</a></li>
<li><a href="08-KanbanTutorial.html">Kanban Boards for Project Management</a></li>
<li><a href="09-GithubCodeReview.html">What Code Review Looks Like in GitHub</a></li>
<li><a href="51-datawrangling.html">Data Wrangling and exploration</a></li>
<li><a href="spreadsheets.html">Details on spreadsheets design</a></li>
<li><a href="16-open.html">Open Science Practices</a></li>
<li><a href="statistics.html">statistics, some details</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-21_Definitions"><p>Content from <a href="21_Definitions.html">Project management, data science</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/21_Definitions.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the benefits of using data science skills in research
project management?</li>
<li>What are common challenges for research teams and research project
management?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li><p>Understanding what project management and data science
entail.</p></li>
<li><p>Knowing the next steps to take after this workshop.</p></li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>In this introduction, we will have an overview of what is meant with
project management. Because we will use data science principles to
tackle the question of open and reproducible research, we will also
define what data science is. Reproducibility is introduced later in the
course, while we expect participants to be quire knowledgable in open
research. One can refer to the <a href="16-open.html">additional
content</a> if openness needs a definition.</p>
<section id="project-management-in-a-research-context"><h2 class="section-heading">Project Management in a research context<a class="anchor" aria-label="anchor" href="#project-management-in-a-research-context"></a>
</h2>
<hr class="half-width">
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What defines a research project ? What makes two projects in a lab
different?</p>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>The answer may vary with domain and lab culture, but the main
components are very similar to what defines one research
publication:</p>
<ul>
<li>The vision or specific research question</li>
<li>The team: different people may be involved inside and outside the
lab</li>
<li>The funding source</li>
<li>The methods used</li>
</ul>
<p>Different types of experiments can be part of the same project.</p>
<p>As lab resources are often pooled between project, this means that
lab and project management are linked, while being independent.</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>The research workflow, also known as the research cycle, usually
starts with a research idea, via a literature search on what is already
known on the subject, to data collection and analysis, to writing,
publishing, and the final assessment of a study. Each of these steps in
a given research project involves aspects of project management that
such as planning, coordination, execution and monitoring of the team,
resources (budgets, materials, equipment), and timeliness according to
schedule.</p>
<div class="section level3">
<h3 id="project-design-and-planning">Project Design and Planning<a class="anchor" aria-label="anchor" href="#project-design-and-planning"></a>
</h3>
<p>It involves defining the project plan which includes the scope,
objectives, milestones, and deliverables within the available
timeline.</p>
<p>This includes the creation of a timetable or Gantt chart usually
created during the grant application.</p>
<p>A great way to visualise the project plan is the Gantt chart, a
horizontal bar chart on a time scale, reflecting all of a projectâs
components, dependencies, and responsibilities. <a href="https://www.ganttproject.biz/" class="external-link">GanttProject</a> and <a href="https://wiki.gnome.org/Apps/Planner/" class="external-link">GNOME Planner</a> are two
open-source project management applications that allow you to visualise
your project plan in a Gantt chart.</p>
</div>
<div class="section level3">
<h3 id="resource-management">Resource Management<a class="anchor" aria-label="anchor" href="#resource-management"></a>
</h3>
<p>It includes managing the budget, allocation of funds to each step
over time, and ensuring that deliverables and objectives will be met
within the available timeframe.</p>
</div>
<div class="section level3">
<h3 id="team-management">Team Management<a class="anchor" aria-label="anchor" href="#team-management"></a>
</h3>
<p>It is the task of making people work together. It includes the
definition of the responsibilities for every team member on specific
steps and aspects of the research project, task assignments, progress
monitoring, and team efficiency measures to thus produce results. It is
core in meeting the objectives and goals of a research project.</p>
<p>Research teams often work remotely with team members being located in
different parts of the world, and therefore need tools that allow for
real-time collaboration, as well as access to process documentation,
files, and data to all team members at any given time.</p>
</div>
<div class="section level3">
<h3 id="data-management">Data Management<a class="anchor" aria-label="anchor" href="#data-management"></a>
</h3>
<p>The core of every research project, including the collection,
organisation, analysis, and secure storing of research data. A <a href="https://en.wikipedia.org/wiki/Data_management_plan" class="external-link">Data
Management Plan</a> (DMP) is increasingly a necessary component of the
project design process and in research proposals, and describes in
detail how and where the collected research will be recorded, stored
securely, and made accessible for analysis and reuse. Furthermore, the
F.A.I.R. Data Principles are equally important to comply with to make
research data human- as well as machine-readability to ensure they are
findable, accessible, interoperable and reusable, and thus FAIR. We will
discuss FAIR data in more detail in episode XX.</p>
</div>
<div class="section level3">
<h3 id="communication-and-reporting">Communication and Reporting<a class="anchor" aria-label="anchor" href="#communication-and-reporting"></a>
</h3>
<p>Effective communication among team members and project stakeholders
(project partners, funders, librarians, publishers, â¦) to keep everyone
updated on the projectâs progress. A key component here is the strategic
documentation and reporting of the methodologies and any adjustments
along the execution of the project. Electronic Lab Notebooks (ELNs)
provide a certain amount of interoperability between systems to automate
some of the documentation processes, thereby making it easier and more
efficient to comply with the FAIR principles.</p>
<ul>
<li>ELNs</li>
<li>Progress reports</li>
</ul>
</div>
<div class="section level3">
<h3 id="risk-management">Risk Management<a class="anchor" aria-label="anchor" href="#risk-management"></a>
</h3>
<p>It is an assessment and documentation of any foreseeable events that
might arise and interfere with the success of the project. and adaption
With a <a href="https://en.wikipedia.org/wiki/SWOT_analysis#" class="external-link">S.W.O.T.
Analysis</a> as part of the project plan, it is possible to describe
strengths and weaknesses of the project idea, and also opportunities and
threats that might impact the project through external factors. A
thorough risk assessment allows you to make contingency plans to address
any challenges or issues should they occur.</p>
</div>
<div class="section level3">
<h3 id="monitoring-and-evaluation">Monitoring and Evaluation<a class="anchor" aria-label="anchor" href="#monitoring-and-evaluation"></a>
</h3>
<p>Monitoring and Evaluation involves ensuring quality control
throughout the execution of the project and the adherence to research
policies and procedures, lessons learned and conclusions that can be
drawn from the results.</p>
</div>
<div class="section level3">
<h3 id="dissemination">Dissemination<a class="anchor" aria-label="anchor" href="#dissemination"></a>
</h3>
<p>Making research results Findable, Accessible, Interoperable, and
Reusable (see again the [F.A.I.R. principles[(<a href="https://www.go-fair.org/fair-principles/" class="external-link uri">https://www.go-fair.org/fair-principles/</a>)) is key to the
dissemination plan of a research project. Datasets and code should be
archived in standardised repositories (see <a href="https://www.re3data.org/" class="external-link">re3data.org</a> for lists of region- and
discipline-specific as well as generalist repositories you can use).</p>
<div id="testimonial1" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial1"></a>
</h3>
<div class="callout-content">
<p>I often manage projects where I do not have decision power. My work
usually starts by making sure decisions are taken and documented. The
documentation is very important, because the implementation of the
decision often requires me to remind people of their decision. Indeed,
during the rest of the project, my role consist in monitoring progresses
and remind people of the objectives, such that specific work is done at
the right time. For instance, a team may well aim for publishing data at
the start of a project, but this objective will not be met unless a
specific workflow is implemented at the start of the project. My work is
then to make sure the researchers are aware of the problems and that
they do not take shortcuts that go against the long term objectives.</p>
<p>As a project manager, one needs to make sure the resources are
sufficient to achieve the goals set in the plan. In research, it often
consists in making sure the people are indeed working toward the goal
and are not investing their time to other projects or other objectives
of the lab. This is particularly important when short term individual
incentives are not aligned with the project long term vision. For
instance, data management is primordial in team science, but, too often,
data collector cannot recognize its importance.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="project-management-tools-overview">Project Management tools overview<a class="anchor" aria-label="anchor" href="#project-management-tools-overview"></a>
</h3>
<p>Several digital project management tools exist that facilitate the
remote coordination and management of research teams with their
projects. Widely used proprietary examples include Trello, Asana,
ClickUp, Notion, and Zoho Projects. Each of these has a different set of
features, and all of them work with a <a href="https://en.wikipedia.org/wiki/Kanban_board" class="external-link">Kanban board</a> for
process documentation.</p>
<p><a href="https://kanboard.org/" class="external-link">Kanboard</a>, <a href="https://wekan.github.io/" class="external-link">WeKan</a>, <a href="https://www.openproject.org" class="external-link">Open project</a>, and <a href="https://taiga.io/" class="external-link">Taiga</a> are examples of open-source project
management software that contain Kanban boards.</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>What project management (digital) tools are you using ?</p>
</div>
</div>
</div>
</div>
</section><section id="what-is-data-science"><h2 class="section-heading">What is data science<a class="anchor" aria-label="anchor" href="#what-is-data-science"></a>
</h2>
<hr class="half-width">
<p>Over the last decade, several tools, methods and training resources
have been developed for early career researchers to learn about and
apply data science skills in biomedicine. This is often referred to as
biomedical data science, with the following definition.</p>
<div id="testimonial2" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial2"></a>
</h3>
<div class="callout-content">
<p>Biosciences and biomedical researchers regularly combine mathematics
and computational methods to interpret experimental data. The term âdata
scienceâ describes expertise associated with taking (usually large) data
sets and annotating, cleaning, organizing, storing, and analyzing them
for the purposes of extracting knowledge. [â¦] The terms âbiomedical data
scienceâ and âbiomedical data scientistâ [â¦] connote activities
associated with the creation and application of methods to new and large
sources of biological and medical data aimed at converting them into
useful information and knowledge. They also connote technical activities
that are data-intensive and require special skills in managing the
large, noisy, and complex data typical of biology and medicine. They may
also imply the application of these technologies in domains where their
collaborators previously have not needed data-intensive computational
approaches.</p>
<p>-- Russ B. Altman and Michael Levitt (2018). <a href="https://www.annualreviews.org/doi/full/10.1146/annurev-bd-01-041718-100001" class="external-link">Annual
Review of Biomedical Data Science</a></p>
</div>
</div>
</div>
<p>In contrast to the definition above (and as will be explained in the
next chapters), we think research which is not data intensive would also
gain in applying data science principles. However, to ensure that data
science approaches are appropriately applied in domain research, such as
in biosciences, there is a need to also engage and educate scientific
group leaders and researchers in project leadership roles on best
practices. Computational methods might indeed be as complex as a neural
network, but even statistical tests and producing figures for a
publication require data science and coding methods.</p>
<p>Researcher use data science skills to apply computation techniques
and reproducible data analyses approaches to their research questions.
In order to apply these tools, researcher first need to understand and
apply the building blocks of data science, especially research data
management, collaborative working and project management.</p>
<figure><img src="fig/handbook-for-application.jpg" alt="Two people with computational expertise holding a giant book towards two other people who conduct lab experiments. The book saya: how to apply data science in biology." width="500" class="figure mx-auto d-block"></figure><p><em>How to apply data science in biology. The Turing Way project
illustration by Scriberia for The Turing Way Community Shared under
CC-BY 4.0 License. Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<div id="testimonial3" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial3"></a>
</h3>
<div class="callout-content">
<p>In some instances, it has been argued that âdata scienceâ simply
rebrands existing fields like statistics or computer science. Our view
is that data science has gained traction as an overarching term due to
increased data availability and complexity; development of computational
methods; advances in computational infrastructure; growing concerns
about scientific rigor and the reproducibility of research findings; and
a recognition that new advances will result from interdisciplinary
research and collaboration. These trends are not unique to data science,
but their integration and consolidation under a single term, however
broad, reflects an understanding of their interconnectedness and is a
real shift in the scientific landscape</p>
<ul>
<li>Goldsmith, J., Sun, Y., Fried, L. P., Wing, J., Miller, G. W., &amp;
Berhane, K. (2021). The Emergence and Future of Public Health Data
Science. Public Health Reviews, 42. doi: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8378512/" class="external-link">10.3389/phrs.2021.1604023</a>
</li>
</ul>
</div>
</div>
</div>
<p>With new technologies supporting the generation of large-scale data
as well as successful applications of data science, Machine Learning
(ML) and Artificial Intelligence (AI) in biomedicine and related fields
have recently shown huge potential to transform the way we conduct
research. Recent groundbreaking research utilising AI technologies in
biomedicine has led to an enormous interest among researchers in data
science, ML and AI approach to extract useful insights from big data,
make new discoveries and address biological questions. As pictured
below, in order to apply these tools, researcher first need to
understand and apply the building blocks of data science, especially
research data management, collaborative working and project
management.</p>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<ul>
<li><p>In what aspects of your projects do you already apply
computational and statistical approaches?</p></li>
<li><p>Do you consider data science relevant for your project? Why/Why
not?</p></li>
<li><p>To what extent do you apply data science practices in your
research projects? (A) Not yet, (B) Sometimes, (C) In most projects, (D)
In every project.</p></li>
</ul>
</div>
</div>
</div>
<p>The <a href="https://github.com/alan-turing-institute/data-training-for-bioscience" class="external-link"><strong>Data
Science for Biomedical Scientists</strong></a> project helps address
this need in training by equipping experimental biomedical scientists
with essential computational skills. In all the resources developed
within this project, we consistently emphasise how computational and
data science approaches can be applied while ensuring reproducibility,
collaboration and transparent reporting.</p>
<div id="testimonial4" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial4"></a>
</h3>
<div class="callout-content">
<p><strong>The goal is to maintain the highest standards of research
practice and integrity</strong>.</p>
</div>
</div>
</div>
<p>In this training material for learning how to manage computational
projects, we discuss essential practices for computational
reproducibility required for carrying out meaningful analyses of
research datasets through data exploration, processing, visualisation
and communication. We present unfamiliar and complex topics from
computation and data science to biologists by providing examples and
recommendations from their fields. The goal is to enable effective
management and sharing of their computational projects. We therefore
encourage you to go through this training material before taking our
second workshop, more focused in AI and Data Science.</p>
</section><section id="project-management-in-open-and-reproducible-projects"><h2 class="section-heading">Project management in open and reproducible projects<a class="anchor" aria-label="anchor" href="#project-management-in-open-and-reproducible-projects"></a>
</h2>
<hr class="half-width">
<p>This course aims at giving an overview of project management
techniques particularly useful for open and reproducible computational
project. This means we will not talk much about risk assessment, and
resource management, but focus on team and data management. This has of
course repercussion in the project design and planning, the
communication and reporting strategy, and monitoring and quality control
is at the core of the code management topic.</p>
<p>In this course, we will also present some software that may help to
share project management tasks and results in a distributed team.</p>
</section><section id="some-references-and-resources"><h2 class="section-heading">Some References and resources<a class="anchor" aria-label="anchor" href="#some-references-and-resources"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="general-guides">General guides<a class="anchor" aria-label="anchor" href="#general-guides"></a>
</h3>
<ul>
<li>A Quick Guide to Organizing Computational Biology Projects Noble WS
(2009) A Quick Guide to Organizing Computational Biology Projects. PLOS
Computational Biology 5(7): e1000424. <a href="https://doi.org/10.1371/journal.pcbi.1000424" class="external-link">https://doi.org/10.1371/journal.pcbi.1000424</a>
</li>
<li>Seddighi, M, Allanson, D, Rothwell, G, Takrouri, K. Study on the use
of a combination of IPython Notebook and an industry-standard package in
educating a CFD course. Comput Appl Eng Educ. 2020; 28: 952â 964. <a href="https://doi.org/10.1002/cae.22273" class="external-link">https://doi.org/10.1002/cae.22273</a>
</li>
<li>Mendez, K.M., Pritchard, L., Reinke, S.N. et al.Â Toward
collaborative open data science in metabolomics using Jupyter Notebooks
and cloud computing. Metabolomics 15, 125 (2019). <a href="https://doi.org/10.1007/s11306-019-1588-0" class="external-link">https://doi.org/10.1007/s11306-019-1588-0</a>
</li>
<li>
<a href="https://goodresearch.dev/setup.html" class="external-link">The Good Research Code
Handbook</a> by Patrick Mineault.</li>
<li>
<a href="https://www.britishecologicalsociety.org/publications/guides-to" class="external-link">Guides
to Better Science by British Ecological Society</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="relevant-turing-way-chapters">Relevant turing way chapters<a class="anchor" aria-label="anchor" href="#relevant-turing-way-chapters"></a>
</h3>
<ul>
<li>
<p><em>The Turing Way</em>. <em>The Turing Way Community. (2021).
The Turing Way: A handbook for reproducible, ethical and collaborative
research (1.0.1). Zenodo. <a href="https://doi.org/10.5281/zenodo.5671094" class="external-link">DOI:
10.5281/zenodo.5671094</a></em></p>
<ul>
<li><a href="https://the-turing-way.netlify.app/project-design/project-repo.html" class="external-link">Guide
to Project Design</a></li>
<li><a href="https://the-turing-way.netlify.app/project-design/project-repo.html" class="external-link">Creating
Project Repositories</a></li>
<li><a href="https://the-turing-way.netlify.app/project-design/filenaming.html" class="external-link">Naming
files, folders and other things</a></li>
<li><a href="https://the-turing-way.netlify.app/reproducible-research/licensing.html" class="external-link">Licensing</a></li>
<li><a href="https://the-turing-way.netlify.app/reproducible-research/ci/ci-options.html" class="external-link">Continuous
Integration</a></li>
<li><a href="https://the-turing-way.netlify.app/reproducible-research/testing.html" class="external-link">Code
Testing</a></li>
<li>
<a href="https://the-turing-way.netlify.app/reproducible-research/code-quality.html" class="external-link">Code
Quality</a> chapter in <em>The Turing Way</em>.</li>
<li><a href="https://the-turing-way.netlify.app/reproducible-research/renv/renv-options.html" class="external-link">Capturing
Computational Environments</a></li>
</ul>
</li>
<li>
<p><a href="https://the-turing-way.netlify.app/reproducible-research/vcs.html" class="external-link">Version
control</a></p>
<ul>
<li><a href="https://the-turing-way.netlify.app/collaboration/github-novice.html" class="external-link">Getting
Started With GitHub</a></li>
<li><a href="https://the-turing-way.netlify.app/reproducible-research/open.html" class="external-link">Open
Science</a></li>
<li><a href="https://the-turing-way.netlify.app/collaboration/new-community.html" class="external-link">Managing
a New Community and Team</a></li>
<li>
<a href="https://the-turing-way.netlify.app/communication/citable.html" class="external-link">Making
Research Objects Citable</a> ## others</li>
</ul>
</li>
<li><p><a href="https://openlifesci.org" class="external-link">Open Life Science training and
Mentoring Programme</a>. <em>Batut, BÃ©rÃ©nice, Yehudi, Yo, Sharan,
Malvika, Tsang, Emmy, &amp; Open Life Science Community. (2021). Open
Life Science - Training and Mentoring programme - Website release
2019-2021 (1.0.0). Zenodo. <a href="https://doi.org/10.5281/zenodo.5636584" class="external-link">DOI:
10.5281/zenodo.5636584</a></em></p></li>
<li><p><a href="https://support.jmir.org/hc/en-us/articles/360003450852-What-is-a-Registered-Report-" class="external-link">What
is a Registered Report?</a> by JMIR, referencing to <a href="https://www.cos.io/initiatives/registered-reports" class="external-link">Registered
Reports by Center for Open Science</a></p></li>
</ul>
</div>
<div class="section level3">
<h3 id="literate-programing-guide">literate programing guide<a class="anchor" aria-label="anchor" href="#literate-programing-guide"></a>
</h3>
<ul>
<li><a href="https://bookdown.org/yihui/rmarkdown" class="external-link">The Definitive
Guide</a></li>
<li><a href="https://jupyter-notebook-beginner-guide.readthedocs.io/en/latest/what_is_jupyter.html" class="external-link">Jupyter/IPython
Notebook Quick Start Guide</a></li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Project management includes resource, team, data, communication and
risk management.</li>
<li>Data science is dealing with large and diverse team, working
remotely while using and re-using code to achieve reproducible
analysis.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-22_contentAndMotivation"><p>Content from <a href="22_contentAndMotivation.html">Course content and motivation</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/22_contentAndMotivation.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How does this training relate to your work?</li>
<li>What are the benefits of using data science skills in research
project management?</li>
<li>What are common challenges for research teams and research project
management?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understanding the scope and structure of this course.</li>
<li>Contextualising computational project management for biomedicine and
life science research.</li>
<li>Knowing the next steps to take after this workshop.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><br><br></p>
<section id="overview-of-this-training-material"><h2 class="section-heading">Overview of this training material<a class="anchor" aria-label="anchor" href="#overview-of-this-training-material"></a>
</h2>
<hr class="half-width">
<p>This chapter gives you an overview of the topics of this course. It
also corresponds to specificity of data science work. They are not
treated in the same order as indicated here.</p>
<p>This program will teach you best practices in data science project
management and how to appyly them to research project. This material
will help you to manage a research project that comprise some <em>online
collaborative working</em>, has a relatively <em>big team, where people
have complementary skills</em>, use <em>coding or programming</em>, as
well as the <em>reuse of code</em>, and last but not least, aim at
producing a <em>reproducible analysis</em>, as is pictured below.</p>
<figure><img src="fig/datasciencespecifics.jpg" alt="Specicity of data science project. Five blocks (working online, large teams whose members have with specialised skills, writing code and re-using code) are placed around a central block where reproducible analysis is written. Data specifics by Julien Colomb CC-BY 4.0 " width="500" class="figure mx-auto d-block"></figure><p>Here we give an short overview of the topics that will be covered in
this course. Note that the course episode split follows a different
logic, and you will find training linked to each five of these data
science specifics in most episodes.</p>
<div id="team-and-data-science" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="team-and-data-science" class="callout-inner">
<h3 class="callout-title">Team and data science<a class="anchor" aria-label="anchor" href="#team-and-data-science"></a>
</h3>
<div class="callout-content">
<p>The principles taught in this course can be applied to any research
project. It will help project manager to work with online tools, with
people with specialised skillsets, and with a project involving some
code writing and code reuse. The goal is to produce a reproducible data
analysis in modern research context.</p>
<p>It requires some specific and some non-specific project management
actions. This course covers a large panel of action to perform before,
during and after the project is conducted.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="online-work">Online work<a class="anchor" aria-label="anchor" href="#online-work"></a>
</h3>
<p>When part of the work happens online, it becomes very easy to loose
track of what documents are where, what tools were used (by whom), and
who is doing what. Discussions are also more difficult to organise and
meetings are more complex to drive. In this course, we will look at
different elements that make this work easier.</p>
<p>Af first, there should be one entry point for the project, where
every team member can find the main documentation as well as links to
other documents and data. This starts with setting
<code>one main URL</code> when setting up the project, as well as using
good readme files and templates. The information needs to be updated
during the project and shared with the whole team.</p>
<p>The use of online project management tool (like kaban boards for todo
list) can also help members of the team to coordinate their work, and
follow their achievements.</p>
</div>
<div class="section level3">
<h3 id="team-science">Team science<a class="anchor" aria-label="anchor" href="#team-science"></a>
</h3>
<p>Because teams can be big, and quite heterogeneous in terms of skills
(especially computer and programming skills), it is important to follow
best practice of team building.</p>
<p>In particular, one should set reasonable goals and milestones for the
project, and document them in the main documentation. It is also
important that every team member knows what his part is, and that the
work is well distributed.</p>
<p>One should make sure every team member is able to use the
communication tools set for the team, and take particular care of the
organisation of meetings. Data and code should be documented (and this
documentation work should be fostered), such that every team member can
follow and reuse the work of the other team members.</p>
</div>
<div class="section level3">
<h3 id="coding">Coding<a class="anchor" aria-label="anchor" href="#coding"></a>
</h3>
<p>When data analysis is done via a programming language, things become
mostly easier, but this facility has some drawbacks, as well as some
effects on data management practice.</p>
<p>First, a data analysis workflow will now start with the computer
reading the raw data. This means that the choice of the data format for
the raw data may change, and that manually gathered data should be
(easily) computer readable. This is particularly important for
spreadsheets, as a lot of time can be saved by designing the spreadsheet
in a tidy format.</p>
<p>Second, statistical analysis and data representation in figures will
now be much easier to perform, making p-hacking and harking practices
very easy to do, even involuntarily. Researcher have to actively make
sure their analysis is not flawed.</p>
<p>Third, making errors in the code may have larger effects than making
error in a manual analysis. However, errors are easier to spot (doing
code reviews and tests) and when the code is corrected, the results are
immediately corrected, too.</p>
</div>
<div class="section level3">
<h3 id="codedata-publication-and-reuse">Code/data publication and reuse<a class="anchor" aria-label="anchor" href="#codedata-publication-and-reuse"></a>
</h3>
<p>Very soon in a research project, writing code consist mostly of
taking code written by someone else and applying it (with some tweaks
sometimes) to one own data. We will look at ways to find relevant code,
make sure it can be trusted, make sure you can legally use it, and ways
to cite it (to give recognition the initial software engineer
deserves).</p>
<p>In addition, code written in the project will probably be reused,
too. We will look into best practices to make this reuse easier, both in
how the code is written (modularity, documentation, tests) and in how
the code is shared (license, repository, version control, release, users
instruction, developers instructions).</p>
</div>
<div class="section level3">
<h3 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h3>
<p>At the core of data science, the analysis reproducibiliy is both a
goal and a huge advantage (in terms of research transparency,
trustworthiness and work efficiency). The use of code is not enough to
get a reproducible analysis, one needs to have access to both the code
and the data used to produce the research result, a concept called
provenance. This may not be trivial, especially if several version of
the code and of the data exist.</p>
<p>In this course, we will have a strong emphasis on version control,
while we will introduce the concepts (and some tools) of provenance, as
well as literate programming (reproducible reports and executable
papers), where the code, the figure and explanatory text are bound in
the same file.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>This course will help you manage team work that involves coding,
using digital tools.</li>
<li>One objective is to produce reproducible analysis and a more
transparent research.</li>
</ul>
</div>
</div>
</div>
<p><br><br><br><br></p>
<hr>
<p><br><br></p>
</div>
</section><section id="your-research-project-is-a-computational-project."><h2 class="section-heading">Your research project is a computational project.<a class="anchor" aria-label="anchor" href="#your-research-project-is-a-computational-project."></a>
</h2>
<hr class="half-width">
<p>As a researcher, you are likely to use some sort of computational
tools to process, analyse, and visualise data. You are also likely to
work on your project with other members of the lab, and the success of
your work may well depend on your interactions with your peers.
<strong>In that sense any research project can be defined as a
collaborative, computational project.</strong></p>
<figure><img src="fig/skill-spectrum.jpg" alt="Researchers represented in a map indicating their journey to understand and apply computational approaches. Some may have just started their journey, some may have come far in the learning and some may have gained proficiency based on their research requirements." width="500" class="figure mx-auto d-block"></figure><p><em>We all may have dfferent research and data science expertise. The
Turing Way project illustration by Scriberia for The Turing Way
Community Shared under CC-BY 4.0 License. Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<div id="why-are-you-here" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="why-are-you-here" class="callout-inner">
<h3 class="callout-title">Why are you here<a class="anchor" aria-label="anchor" href="#why-are-you-here"></a>
</h3>
<div class="callout-content">
<p>Discuss why you/learners are taking this course, what are the
expectations. Does the expectations align with the relevance of data
science and content of this course?</p>
</div>
</div>
</div>
<p>Contents of this training material introduces methods and concepts to
manage individuals and teams working on any computational project, which
in the current era is literally all research projects. It is
<em>not</em> about learning how to write code, but building a
foundational understanding for computational methods that could be
applied to your research. Furthermore, this training will provide
guidance for facilitating collaboration and data analysis using tools
like research data management, version control or code review.</p>
<p>We acknowledge the data science knowledge will vary. Nonetheless, we
believe that the data science skills you will learn in this training
will make your research process better. In the following sections, we
will detail what we mean by âbetterâ.</p>
</section><section id="how-data-science-will-improve-your-research"><h2 class="section-heading">How data science will improve your research ?<a class="anchor" aria-label="anchor" href="#how-data-science-will-improve-your-research"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/healthy-research-tree.jpg" alt="Researchers pour water on a tree, the water represents data science, the tree is the research." width="500" class="figure mx-auto d-block"></figure><p><em>Data science makes research flourish. The Turing Way project
illustration by Scriberia for The Turing Way Community Shared under
CC-BY 4.0 License. Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<div id="it-is-mostly-about-being-efficient" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="it-is-mostly-about-being-efficient" class="callout-inner">
<h3 class="callout-title">It is mostly about being efficient<a class="anchor" aria-label="anchor" href="#it-is-mostly-about-being-efficient"></a>
</h3>
<div class="callout-content">
<p>Data science brings some structure in how data is collected,
processed and analysed, making it easier to collaborate on a project, to
publish extra research outputs and leveraging some extra potential your
data may have. In the past, it helped me drive new hypotheses, detect
problems with the research design early, and reduce the sample size
needed to drive a solid conclusion. Eventually, it made my research more
robust and trustworthy. But in the end, my real motivation is
efficiency: very soon, the time I invested in learning and applying data
science in my research was recovered multiple times when a manuscript
had to be written (and re-written).</p>
</div>
</div>
</div>
<p>There are different ways to organise the different foreseen
improvement, we decided here to start with improvement in the final
result, improvement in the research process, and finally aspects of
community building.</p>
<div class="section level3">
<h3 id="using-code-for-nicer-paper">Using code for nicer paper<a class="anchor" aria-label="anchor" href="#using-code-for-nicer-paper"></a>
</h3>
<div class="section level4">
<h4 id="powerful-statistics">Powerful statistics<a class="anchor" aria-label="anchor" href="#powerful-statistics"></a>
</h4>
<p>The most advanced statistical methods (like machine learning) are
first developed in programming environment, and they are often difficult
or impossible to implement in statistical software. In addition, some of
the less advanced statistical methods requires intensive data processing
that makes it very difficult to apply outside a coding environment.</p>
<div class="section level5">
<h5 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h5>
<ol style="list-style-type: decimal">
<li>
<a href="https://logarithmic.net/langevitour/2022-useR/#2" class="external-link">logarithmic.net/langevitour/2022-useR/#2</a>
to explore orthonormal projections of high dimensional data.</li>
<li>Results of water maze behavioural tests are better analysed using a
survival analysis than an <a href="https://en.wikipedia.org/wiki/Analysis_of_variance" class="external-link">Analysis of
variance (ANOVA)</a>. However, the data obtained via video analysis
software is often not fitted for that analysis and needs to be
transformed. Doing transformations by hand is time consuming and is
likely to introduce errors.</li>
</ol>
<p><strong>Exampleâs references</strong></p>
<ul>
<li>See <em>Jahn-Eimermacher, A., Lasarzik, I., &amp; Raber, J. (2011).
Statistical analysis of latency outcomes in behavioral experiments.
Behavioural Brain Research, 221(1), 271â275. doi: <a href="https://www.sciencedirect.com/science/article/abs/pii/S0166432811001847?via%3Dihub" class="external-link">10.1016/j.bbr.2011.03.007</a></em>
</li>
<li>For water maze experiment analysis, see: Maugard, M., Doux, C.,
&amp; Bonvento, G. (2019). A new statistical method to analyze Morris
Water Maze data using Dirichlet distribution. F1000Research, 8(1601),
1601. doi: <a href="https://doi.org/10.12688/f1000research.20072.2" class="external-link">10.12688/f1000research.20072.2</a>.</li>
</ul>
</div>
</div>
<div class="section level4">
<h4 id="informative-and-inclusive-figures">Informative (and inclusive) figures<a class="anchor" aria-label="anchor" href="#informative-and-inclusive-figures"></a>
</h4>
<p>Once you start using code for analysing your data, it will become
much easier to do complex and informative visualisation. This often
includes way to visualise and label single data points, or use
visualisation on several dimensions (producing moving gifs of a 3D
scatterplot for instance).</p>
<p>One can also automate the figure design choice, so that all figures
look similar. Similarly, the production of several version of the same
figure is very easy. For example, one can use different color pallette,
one using the palette usually used in the field (the one your supervisor
wants to see), and one for color-blind readers.</p>
<div id="single-flights-from-different-bees." class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="single-flights-from-different-bees." class="callout-inner">
<h3 class="callout-title">Single flights from different bees.<a class="anchor" aria-label="anchor" href="#single-flights-from-different-bees."></a>
</h3>
<div class="callout-content">
<p>See a good example of data representation in differen format single
flights from different bees shown in supplemnentary data: <em>Menzel,
R., Greggers, U., Smith, A., Berger, S., Brandt, R., Brunke, S., â¦Watzl,
S. (2005). Honey bees navigate according to a map-like spatial memory.
Proceedings of the National Academy of Sciences of the United States of
America, 102(8), 3040. doi: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC549458/" class="external-link">10.1073/pnas.0408550102</a></em></p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="reproducible-analysis">Reproducible analysis<a class="anchor" aria-label="anchor" href="#reproducible-analysis"></a>
</h4>
<p>Good scientific practices are aiming at securing the robustness and
reproducibility of the scientific endeavour. As a researcher, assuring
computational reproducibility of your results is a relatively easy step
in making your research more robust.</p>
<figure><img src="fig/ReproducibleJourney.jpg" alt="Shows a landscape with different checkpoints fpr data, code, tools and result each of which require reproducible practices. There is a woman explaining her reproducibility journey to help new people start their journey" width="500" class="figure mx-auto d-block"></figure><p><em>What to expect in your reproducibility journey. The Turing Way
project illustration by Scriberia for The Turing Way Community Shared
under CC-BY 4.0 License. Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<p>The reproducibility of an experiment not only requires a detailed
description of the methods and reagents used, but also a detailed
description of the analysis performed. The ultimate description of the
analysis is to provide all elements necessary for reproducing the
analysis (computational reproducibility). This includes the data and the
code used to analyse it (in a form that can be reused in a different
computational environment).</p>
<div id="callout4" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout4"></a>
</h3>
<div class="callout-content">
<p>In practice, sometimes one may not be able to provide all elements
openly (for instance, some medical data cannot be shared openly for
privacy reasons) for everyone to be able to reproduce all the results.
But co-workers (and maybe reviewers) should be able to reproduce the
analysis (e.g., on anonymized data).</p>
</div>
</div>
</div>
<p>The emergence of reproducible reports is another aspect of computer
reproducibility. Literate programming using Jupyter notebooks,
Rmarkdown, stencila or quarto (tools that can usually use Python, R, or
Julia) is indeed growing in popularity. These tools allow you to show
data and analysis side by side, with written explanations and
interactive visualisations. These outputs can not only be used as blog
posts, or lab reports, but can also be published as an enhanced
publication, a concept called executable research articles: <a href="https://gmaciocci.medium.com/list/the-evolution-of-executable-research-articles-823e42a9fa60" class="external-link">https://gmaciocci.medium.com/list/the-evolution-of-executable-research-articles-823e42a9fa60</a></p>
</div>
<div class="section level4">
<h4 id="diversifying-research-outputs">Diversifying research outputs<a class="anchor" aria-label="anchor" href="#diversifying-research-outputs"></a>
</h4>
<p>While the main recognition currency in academia is still (first)
authorship in peer reviewed publication, new scientometrics are
developed to recognize publication of other research outputs. In
particular, datasets and software publication are officially reviewed in
the evaluation of certain grant, for example for the Marie-curie
european program. Data science principles will make it easier to publish
datasets, software, reagents or hardware you are anyway producing during
the research process.</p>
<div id="callout5" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout5"></a>
</h3>
<div class="callout-content">
<p>By publishing datasets and code, you will not only help other
researchers, but gain extra recognition for your work. However, open
data and open code requires a specific documentation, which we will
touch upon in this training.</p>
</div>
</div>
</div>
<p>Computational tools you produce in your lab can be released as open
source software and credit will be given globally. This may also be true
for hardware you design (this aspect will not be discussed in this
training) or datasets you collect.</p>
</div>
</div>
<div class="section level3">
<h3 id="improved-research-process">Improved Research Process<a class="anchor" aria-label="anchor" href="#improved-research-process"></a>
</h3>
<div class="section level4">
<h4 id="data-quality">Data quality<a class="anchor" aria-label="anchor" href="#data-quality"></a>
</h4>
<p>Did you know that manually copy-pasting data is one of the primary
source of data corruption ? Combining datasets or processing data (such
as cleaning or transforming into different formats), can therefore be
best achieved using code. The process is not only safer, but it is often
faster.</p>
<p>Another underevaluated issue is the amount of data you will collect.
The more data you have the more sophisticated tools and workflows you
may need. It is also more likely that your data (or code) gets
corrupted, mixed up, outdated, or lost. In particular, when something
goes wrong during the experiment, code might be used to create warnings,
so that the setup or protocol may be modified on the fly or between
sessions.</p>
</div>
<div class="section level4">
<h4 id="reproducibility-and-automation">Reproducibility and automation<a class="anchor" aria-label="anchor" href="#reproducibility-and-automation"></a>
</h4>
<p>While we already mentioned the advantage of reproducibile analysis
for the quality of the research, we did not mention how useful this is
during the research. With reproducible analysis, it is effortless to run
a new dataset in the analysis workflow, it becomes possible to explore
the data at a single experiment level, which may enable new hypotheses,
or spot issues that were not foreseen in the experimental design. It
makes also certain that difference in the figures are due to difference
in the dataset, not any manual processing of the data one may have
forgotten to docuzment.</p>
</div>
<div class="section level4">
<h4 id="collaborative-working">Collaborative working<a class="anchor" aria-label="anchor" href="#collaborative-working"></a>
</h4>
<p>Within science teams, group work is critical for experimental design
and implementation. In addition, there are rapid developments in how
scientific results and methods are shared, and collaborations have never
been more global or rapid. This means that several people will likely be
working with the same data files.</p>
<p>Data science allows for the management of how one or multiple people
work on the same project (as well as the same code). It requires
different skillsets than those taught in traditional science courses
<em>or</em> a typical coding class.</p>
<div id="who-can-add-to-your-research" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="who-can-add-to-your-research" class="callout-inner">
<h3 class="callout-title">Who can add to your research?<a class="anchor" aria-label="anchor" href="#who-can-add-to-your-research"></a>
</h3>
<div class="callout-content">
<p>Facilitating communication and sharing will make it easier for your
colleagues to help you. Can you think of people who can help you in your
research, directly in your lab or at your institution ? Would it help
for them to have access to your data? How could they participate, and
how can you give them credit?</p>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Needs from the future you
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>It is very interesting to consider your future self as one
collaborator in your project. Anything you may forget in the next three
to five years should be documented, if you want your future self to be
able to (re-)analyse the data you are collecting. Indeed, the advantage
of working collaboratively in a project can indeed be translated
directly in a project you drive mostly alone.</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="efficiency">Efficiency<a class="anchor" aria-label="anchor" href="#efficiency"></a>
</h4>
<div id="callout6" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout6"></a>
</h3>
<div class="callout-content">
<p>The time invested in your data and code will be paid multiple times
by the efficiency improvement in your workflow, if that investment is
done early in the project. Because one can consider your past self as
one of your collaborator, the advantage of working collaboratively in a
project can indeed be translated directly in a project you drive mostly
alone.</p>
</div>
</div>
</div>
<p>At this point, you may be convinced that the extra work of designing
your project using data science principles will be worth it. But here
comes the best argument of all: <strong>in the end you will save
time.</strong> Early time saving comes because your future self and
collaborators will be able to find all your data, reuse and modify your
code, and understand your research faster.</p>
<p>This applies directly to the example of working on article revisions
- will you still remember all the analysis details and data nuances when
your papers comes back with a request for major changes? For instance,
if a colleague cannot find what data goes with which figures, there are
high chances that you will also be unable to find it three years from
now. In addition, itt is not uncommon to modify the design of the
figures multiple times (sometimes back and forth), often modifying all
figures at once.</p>
<div id="redoing-all-figures-in-minutes" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div id="redoing-all-figures-in-minutes" class="callout-inner">
<h3 class="callout-title">Redoing all figures in minutes<a class="anchor" aria-label="anchor" href="#redoing-all-figures-in-minutes"></a>
</h3>
<div class="callout-content">
<p>Once a reviewer ask me to overlay individual data points onto all our
5 boxplots figures. The project was an old one, and I had not touched
the data for years. Finding the right data and redo the all 5 figures
would usually take ages using SPSS or excel. But since I used code, I
had all figures 15 minutes later. (Note, after seeing the new figures,
the reviewer agreed that the original version was better).</p>
</div>
</div>
</div>
<p>Later on in the project, community advantages are coming in. Data and
code reusability is not only a mark of research transparency and
robustness, it also means you can reuse your own code and data. It also
means you can reuse code and data produced by other researchers.</p>
<p>The snow ball effect may be huge, and the objective of this course is
to allow you to do <strong>better science in less time</strong></p>
<div id="invest-in-data-science" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="invest-in-data-science" class="callout-inner">
<h3 class="callout-title">Invest in data science<a class="anchor" aria-label="anchor" href="#invest-in-data-science"></a>
</h3>
<div class="callout-content">
<p>As an example it was estimated that research data management takes
about 5% of your time, on the other hand, time lost due to poor data
management is estimated to be 15%. See reference: <em>Lowndes, J. S. S.,
Best, B. D., Scarborough, C., Afflerbach, J. C., Frazier, M. R., OâHara,
C. C., Halpern, B. S. (2017). Our path to better science in less time
using open data science tools. Nature Ecology &amp; Evolution, 1(0160),
1â7. doi: 10.1038/s41559-017-0160</em></p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="team-and-community-building">Team and community building<a class="anchor" aria-label="anchor" href="#team-and-community-building"></a>
</h3>
<figure><img src="fig/research-foundation.jpg" alt="A house representing machine learing and AI is set upon bricks that one person is sliding below the house. On the bricks, we can read data science principles like open science, backups, reproducibiliy, and FAIR principles." width="500" class="figure mx-auto d-block"></figure><p><em>Data science foundations. The Turing Way project illustration by
Scriberia for The Turing Way Community Shared under CC-BY 4.0 License.
Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<p>Data science tools will make it easier not only to collaborate with
researchers in your lab, but also with researchers outside of your lab,
or even with non-researchers (citizen science or software
professionals). These may bring valuable expertise in the project. Being
part of a collaborative community will also create impact beyond
citations and papers, something which starts to be valued by funding
agencies, and which make research more fun, valued and interesting.</p>
<p>We may also add to the pot that creating a network around your
research is a critical aspect of building a career in academia. Being
known as a good and skilled collaborator can open doors to many
opportunities.</p>
</div>
</section><section id="a-journey-starts"><h2 class="section-heading">A journey starts<a class="anchor" aria-label="anchor" href="#a-journey-starts"></a>
</h2>
<hr class="half-width">
<blockquote>
<p>You step into the Road, and if you donât keep your feet, there is no
knowing where you might be swept off to.</p>
<p>J.R.R. Tolkien, The Lord of the Rings</p>
</blockquote>
<p>This training will give you some starting points, but implementating
data science principles is a long and always renewed process. But you do
not need to do it all at once, and you do not need to do it alone.</p>
<p>After the training, do not hesitate to join (or create) a community
of like-minded researchers where you live (there are always some if you
look). In addition, there may be people at your institution whose job is
to help you. Look for data steward or data managers, research data
engineers, IT support, open science bÃ¼ros at your institution and be
pro-active contacting them. There are also almost endless online
resources and helpful communities. For instance, <em>The Turing Way</em>
guide for data science and research provides several detailed chapters
to cover topics across reproducibility, project design, collaboration,
communication, research ethics and community building.</p>
<figure><img src="fig/ttw-welcome.jpg" alt="drawing" width="400" class="figure mx-auto d-block"></figure><p><em>The Turing Way project illustration by Scriberia for The Turing
Way Community Shared under CC-BY 4.0 License. Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ul>
<li><p>The course provides an introduction to Managing Open and
Reproducible Computational Projects to senior biomedical researchers, in
order to equip them with tools and techniques to generate and document
and manage complex reproducible computational projects.</p></li>
<li><p>Without going into underlying technical details, the contents
provide a general overview and present selected relevant biomedical case
studies.</p></li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-23_Ensuring-reproducibility"><p>Content from <a href="23_Ensuring-reproducibility.html">Ensuring reproducibility</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/23_Ensuring-reproducibility.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li><p>How to build reproducible analysis?</p></li>
<li><p>how to deal with dependencies?</p></li>
<li><p>What is version control?</p></li>
<li><p>Why using git ?</p></li>
<li><p>How is version control system relevant for biomedical
research?</p></li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the different concepts related to reproducibility.</li>
<li>Learn about methods to capture reproducible research
environments.</li>
<li>Understand the concept of provenance and version control.</li>
<li>Understand the power and pitfalls of Git</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="what-is-computer-reproducibility"><h2 class="section-heading">What is computer reproducibility<a class="anchor" aria-label="anchor" href="#what-is-computer-reproducibility"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/reproducible-matrix.jpg" alt="A matrix showing data and analysis in two axis and iterating that reproducibility is when same analysis is applied to same data it gives same result." class="figure mx-auto d-block"></figure><p><em>Documentation as a guiding light for people who may feel lost
otherwise. The Turing Way project illustration by Scriberia for The
Turing Way Community Shared under CC-BY 4.0 License. Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<p>The different dimensions of reproducible research described in the
matrix above have the following definitions directy taken from <em>The
Turing Way</em> Guide to Reproducible Research (see the <a href="https://the-turing-way.netlify.app/reproducible-research/overview.html" class="external-link">oveview
chapter</a>):</p>
<ul>
<li>
<strong>Reproducible:</strong> A result is reproducible when the
<em>same</em> analysis steps performed on the <em>same</em> dataset
consistently produces the <em>same</em> answer.</li>
<li>
<strong>Replicable:</strong> A result is replicable when the
<em>same</em> analysis performed on <em>different</em> datasets produces
qualitatively similar answers.</li>
<li>
<strong>Robust:</strong> A result is robust when the <em>same</em>
dataset is subjected to <em>different</em> analysis workflows to answer
the same research question (for example one pipeline written in R and
another written in Python) and a qualitatively similar or identical
answer is produced. Robust results show that the work is not dependent
on the specificities of the programming language chosen to perform the
analysis.</li>
<li>
<strong>Generalisable:</strong> Combining replicable and robust
findings allow us to form generalisable results. Note that running an
analysis on a different software implementation and with a different
dataset does not provide <em>generalised</em> results. There will be
many more steps to know how well the work applies to all the different
aspects of the research question. Generalisation is an important step
towards understanding that the result is not dependent on a particular
dataset nor a particular version of the analysis pipeline.</li>
</ul></section><section id="provenance-and-reproducibility"><h2 class="section-heading">Provenance and reproducibility<a class="anchor" aria-label="anchor" href="#provenance-and-reproducibility"></a>
</h2>
<hr class="half-width">
<p>In order to make results reproducible, one needs to access and reuse
the data and the code used to produce the analysis. Since both code and
data may evolve over time (errors may be found, variables may be
changed), it is important to link the figure created with a specific
version of the code and the data, this is called keeping provenance
information.</p>
<figure><img src="fig/data-provenance.jpg" alt="A figure explaining the concept of provenance: Silos labeled raw data, preprocessed data, analysis, and paper are connected and points representing data are flowing from the first to the last silo." class="figure mx-auto d-block"></figure></section><section id="long-term-reproducibility"><h2 class="section-heading">Long term reproducibility<a class="anchor" aria-label="anchor" href="#long-term-reproducibility"></a>
</h2>
<hr class="half-width">
<p>Both data and software should be in a form where the analysis can run
on a different hardware. Because software evolve quite rapidly, it may
be difficult to have the analysis run years after its creation, because
the dependencies will have evolve and may not be compatible with the old
code anymore.</p>
<p>It is therefore important to know what version of the dependencies
were used. In practices, there are specific tools to record
dependencies, and for complex code, it is preferable to use countainers.
Containers are specific piece of software that âcontainâ all the
information and dependences needed to run a code.</p>
<p>Alternatively, one can aim at robustness and document the analysis
enough such that adapting the code to a new computer environment is
easy. This may be a easier workflow for simple analysis.</p>
<p>Thinking about which software, tools and platforms to use will
greatly affect how you analyse and process data, as well as how you
share your results for computational reproducibility. The idea is to
facilitate others in recreating the setup process necessary to reproduce
your research.</p>
<div id="some-tools-that-can-be-used-to-enable-these-are-the-following" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="some-tools-that-can-be-used-to-enable-these-are-the-following" class="callout-inner">
<h3 class="callout-title">Some tools that can be used to enable these
are the following:<a class="anchor" aria-label="anchor" href="#some-tools-that-can-be-used-to-enable-these-are-the-following"></a>
</h3>
<div class="callout-content">
<ul>
<li>
<strong>Dependency managers</strong> such as <a href="https://docs.conda.io/en/latest/" class="external-link">Conda</a> keep dependencies
updated and make sure the same version of dependencies used in the
development environments are also used when reproducing a result.</li>
<li>
<strong>Containers</strong> such as <a href="https://www.docker.com/" class="external-link">Docker</a> is a way to create
computational environments with configurations required for developing,
testing and using research software isolated/independent from other
applications.</li>
<li>
<strong>Literate Programming</strong> using <a href="https://jupyter.org/" class="external-link">Jupyter Notebook</a> is an extremely
powerful way to use a web-based online interactive computing environment
to execute code and script while adding notes and additional information
about the application.</li>
</ul>
</div>
</div>
</div>
<div class="section level3">
<h3 id="reproducible-research-environment">Reproducible Research Environment<a class="anchor" aria-label="anchor" href="#reproducible-research-environment"></a>
</h3>
<p>Researchersâ working environments evolve as they update software,
install new software, and move to different computers. If the project
environment is not captured and the researchers need to return to their
project after months or years (as is common in research), they will be
unable to do so confidently.</p>
<p>This includes features of hardware (such as the numbers of cores in
any CPUs) and features of the software (such as the operating system,
programming languages, supporting packages, other pieces of installed
software, along with their versions and configurations).</p>
<figure><img src="https://the-turing-way.netlify.app/_images/computational-environments.jpg" alt="Ways of capturing computational environments" class="figure mx-auto d-block"></figure><p>To learn more about how to create a reproducible environment, the
chapter on <a href="https://the-turing-way.netlify.app/reproducible-research/renv.html" class="external-link">Reproducible
Environments</a> in <em>The Turing way</em> is a good place to
start.</p>
<p>There are several ways of capturing computational environments. The
major ones covered in this chapter will be Package Management Systems,
Binder, Virtual Machines, and Containers. Each has its pros and cons,
and the most appropriate option for you will depend on the nature of
your project. They can be broadly split into two categories: those that
capture only the software and its versions used in an environment
(Package Management Systems), and those that replicate an entire
computational environment - including the operating system and
customised settings (Virtual Machines and Containers).</p>
<p>Another way these can be split is by how the reproduced research is
presented to the reproducer. Using <strong>Binder</strong> or a
<strong>Virtual Machine</strong> creates a much more graphical, GUI-type
result. In contrast, the outputs of <strong>Containers</strong> and
<strong>Package Management</strong> Systems are more easily interacted
with via the command line. Please read more about each of these concepts
and their practice use, please visit <a href="https://the-turing-way.netlify.app/reproducible-research/renv/renv-options.html" class="external-link">Capturing
Computational Environments</a> in <em>The Turing Way</em>.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Provenance information and software dependencies management are
needed to reproduce an analysis on the mid-term (3 years)</li>
<li>Information about the version of the code and of the data used to
create an analysis should be captured.</li>
</ul>
</div>
</div>
</div>
<p><br><br><br><br></p>
<hr>
<p><br><br></p>
</div>
</section><section id="maintaining-history-through-version-control"><h2 class="section-heading">Maintaining History through Version Control<a class="anchor" aria-label="anchor" href="#maintaining-history-through-version-control"></a>
</h2>
<hr class="half-width">
<figure><img src="https://zenodo.org/api/iiif/v2/0c0188d3-d03c-4830-a6e3-00405f5c22fa:df931888-09af-4eac-945f-0c208be0c26b:ProjectHistory.jpg/full/750,/0/default.jpg" alt="Contrast in project history management. On the left - choosing between ambiguosly named files. On the right - picking between successive versions (from V1 to V6)." class="figure mx-auto d-block"></figure><p><em>Version control allows tracking of history and go back to
different versions as needed. The Turing Way project illustration by
Scriberia for The Turing Way Community Shared under CC-BY 4.0 License.
Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<p>Practices and recommendations described in this lesson are applicable
to all areas of biomedical research. What can be considered slightly
different in computational projects is that every object required to
carry out the research exists in digital form: starting from research
workflow, data, software, analysis processes, outcomes as well as how
researchers involved in the project communicate with each other. This
means that research objects and workflows can be organised and
maintained without losing its provenance or missing knowledge of how
each of these objects is connected in the context of your project.</p>
</section><section id="versioning-every-research-object"><h2 class="section-heading">Versioning Every Research Object<a class="anchor" aria-label="anchor" href="#versioning-every-research-object"></a>
</h2>
<hr class="half-width">
<p>Management of changes or revisions to any type of information made in
a file or project is called <em>versioning</em>. Version Control Systems
(VCS) are platforms and technical tools that allow the recording of any
changes made over time in a file or research object. All collaborators
on a research project can track file history, review changes made, track
and fix errors when they appear, and revert or go back to earlier
versions. It is therefore easy to monitor contributions and give
appropriate credit to all contributors.</p>
<p>Different VCS can be used through a program with web browser-based
applications (such as <a href="https://docs.google.com/" class="external-link">Google Docs</a>
for documents) and more dynamically for code and all kinds of data
through command-line tools (such as <a href="https://en.wikipedia.org/wiki/Git" class="external-link">Git</a>) and their integration
into the graphical user interface (<a href="https://code.visualstudio.com/" class="external-link">Visual Studio Code</a> editor, <a href="https://git-scm.com/downloads/guis" class="external-link">Git-gui</a> and <a href="https://www.gitkraken.com/" class="external-link">gitkraken</a>). The practice of
versioning is particularly important to allow non-linear or branched
development of different parts of the project, testing a new feature,
debugging and error or reusing code from one project to different data
by different contributors.</p>
<p><a href="https://about.gitlab.com/" class="external-link">GitLab</a>, <a href="https://github.com/" class="external-link">GitHub</a>, or <a href="https://bitbucket.org" class="external-link">BitBucket</a> are online platforms that
allow version-controlled projects online and allow multiple
collaborators to participate. Different members can download a copy of
the online repository (most recent version), make changes by adding
their contributions locally on their computer and push the changes to
GitLab/GitHub/BitBucket (a new version!) allowing others to build on the
new development.</p>
<p>Read <em><a href="https://towardsdatascience.com/all-you-need-to-know-about-git-github-gitlab-24a853422ff3" class="external-link">All
you need to know about Git, GitHub &amp; GitLab</a></em> on <a href="https://the-turing-way.netlify.app/reproducible-research/vcs.html" class="external-link">Towards
Data Science and version control</a> in <em>The Turing way</em> for more
details on workflow, technical details of using git and versioning large
datasets.</p>
</section><section id="basics-of-version-control"><h2 class="section-heading">Basics of Version Control<a class="anchor" aria-label="anchor" href="#basics-of-version-control"></a>
</h2>
<hr class="half-width">
<p>We have all seen a simple file versioning approach where different
versions of a file are stored with a different name. Often, files are
getting a number suffix that grows with the version number (v001 to v999
for example). When many version exists, it is useful to store old
version in a subfolder and keep a record of what has changed in each
version. Git is a tool that automates that process. As explained below,
Git uses repositories to handle collaborative working, but is not a
real-time collaborative tool.</p>
<figure><img src="fig/manualversioncontrol.jpg" alt='A figure showing manual versus GIT version control. On the left, we have one folder with a file named thesis_latest, a spreadsheet and one subfolder; on the right one folder with a file named thesis and a hidden folder called ".git". Below a screenshot of the spreadsheet on the left with some comments for each version saved in the subfolder; on the right a view of a Git software showing the same comments entered via Git, and a view of the difference added in one of the version. Figure by Julien Colomb, CC-BY 4.0.' class="figure mx-auto d-block"><figcaption>A figure showing manual (left) versus GIT (right)
version control.</figcaption></figure><p>Tools such as Google Drive and Microsoft Teams offer platforms to
share and edit files with others in real-time, collaboratively. More
sophisticated VCS exists within tools like <a href="https://docs.google.com/" class="external-link">Google docs</a> or <a href="https://hackmd.io/" class="external-link">HackMD</a>. These allow collaborators to
update files while storing each version in its version history (we will
discuss this in detail). Advanced VCS such as <a href="https://git-scm.com/" class="external-link">Git</a> and <a href="https://www.mercurial-scm.org/" class="external-link">Mercurial</a> provide much more
powerful tools to maintain versions in local files and share them with
others.</p>
<p>Web-based Git repository hosting services like <a href="https://about.gitlab.com/" class="external-link">GitLab</a> and <a href="https://github.com/" class="external-link">GitHub</a> facilitate online collaborations
in research projects by making changes available online more frequently,
as well as enabling participation within a common platform from
colleagues who donât code. With the help of comments and commit
messages, each version can be annotated with the changes it contains
compared to the previous versions. This is helpful when we share our
analysis (not only data), and make it auditable and reproducible - in
accordance with good scientific practice. In the next chapters, we will
discuss version control for different research objects.</p>
<p>You can read more details in <a href="https://the-turing-way.netlify.app/reproducible-research/vcs.html" class="external-link">Version
Control</a> and <a href="https://the-turing-way.netlify.app/collaboration/github-novice.html" class="external-link">Getting
Started With GitHub</a> chapters in <em>The Turing Way</em>.</p>
</section><section id="git-data-and-provenance"><h2 class="section-heading">Git, data and provenance<a class="anchor" aria-label="anchor" href="#git-data-and-provenance"></a>
</h2>
<hr class="half-width">
<p>As explained in <a href="https://the-turing-way.netlify.app/index.html" class="external-link">The Turing Way
ebook</a>, Git does not work well for large or many files, and therefore
are not well suited for large datasets, presentations or posters. In
order to keep these under Git version control, it is often necessary to
use e.g.Â <a href="https://git-annex.branchable.com/" class="external-link">Git-annex</a> and
submodules. <a href="https://www.datalad.org/" class="external-link">DataLad</a> is a great
tool to use on top of git to manage and collaborate on large
datasets.</p>
</section><section id="git-and-contribution-record"><h2 class="section-heading">Git and contribution record<a class="anchor" aria-label="anchor" href="#git-and-contribution-record"></a>
</h2>
<hr class="half-width">
<p>By using Git for version control, the system will record what was
changed by whom in a repository. This may facilitate the record of
contributions, but one needs to be aware of the technical bias it
creates (importance of contributions cannot be recorded by size or
number of commits, less technical contributors will have less Git
contribution). The Contributor Roles Taxonomy <a href="https://credit.niso.org/" class="external-link">CRediT</a> is a good way to specify and
acknowledge contribution by the people involved and can be easily used
with the web-based app and R package <a href="https://tenzing.club" class="external-link">tenzing</a> .</p>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ul>
<li><p><a href="https://git-scm.com/" class="external-link">Git</a> is a software developed to
facilitate version control of text files</p></li>
<li><p>Version controlled repositories help record different
contributions and contributor information openly.</p></li>
</ul>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<ul>
<li><p>Which of the reproducible research principles are you already
implementing in your work and how?</p></li>
<li><p>How do you currently version control your projects? What changes
will you introduce to ensure fair credit and citation of your and your
teamâs work?</p></li>
<li><p>What are other measures to ensure reproducibility in your
research documentation and that of your team?â</p></li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-24_SettingUp"><p>Content from <a href="24_SettingUp.html">Managing project start and collaborations</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/24_SettingUp.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li><p>How to set up a computational project?</p></li>
<li><p>What main concerns and challenges exist and how to address
them?</p></li>
<li><p>How to create a project repository for sharing, collaboration and
an intention to release?</p></li>
<li><p>What extra tools may be used ?</p></li>
<li><p>How to manage and oversee tasks and track progress of your
projects?</p></li>
<li><p>How collaborative practices help ensure code quality, testing and
reuse?</p></li>
<li><p>What is literate programming and how does it help with early
communication, testing and collaboration?</p></li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li><p>Describe best practices for setting a project repository</p></li>
<li><p>Build a basis for collaboration and co-creation in team
projects</p></li>
<li><p>Plan computational reproducibility and project management
practices</p></li>
<li><p>Make it easy for each contributor to participate, contribute and
be recognised for their work</p></li>
<li><p>Demonstrate GitHub Project Board to enable project
management.</p></li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="setting-up-a-project"><h2 class="section-heading">Setting up a Project<a class="anchor" aria-label="anchor" href="#setting-up-a-project"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/research-cycle.jpg" alt="The research process is represented as a perpetual cycle of generating research ideas, performing data planning and design, data collection, and data processing and analysis, publishing, preserving and hence, allowing re-use of data." class="figure mx-auto d-block"></figure><p><em>Research Lifecycle. The Turing Way project illustration by
Scriberia for The Turing Way Community Shared under CC-BY 4.0 License.
Zenodo. <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">http://doi.org/10.5281/zenodo.3332807</a></em></p>
<p>A research project starts right with a research idea. We start by
communicating that with other researchs in our team. Then come the
following steps:</p>
<ol style="list-style-type: decimal">
<li>planning and designing the research work</li>
<li>describing the research protocols</li>
<li>deciding how data will be collected</li>
<li>selecting methods and practices for processing and wrangling
data</li>
<li>conducting our studies and analysis</li>
<li>publishing all the research objects so everybody can access it</li>
<li>archiving it to ensure that our research is reusable, meaning, that
someone else can go through this whole process of reproducing or
building upon our work.</li>
</ol>
<p>Each of these steps is important for every single researcher,
irrespective of their roles in the project. They are usually not taken
one after the other, but concomitantly, it is therefore important to
plan and take decisions for each of them before the project starts.
However, a project lead (such as Principal Investigators, managers and
supervisors) have an added responsibility to set up the project in a way
that ensures that all members of their research team can work together
efficiently at all stages of the project.</p>
<p>With an overarching goal to maintain research integrity and ethical
practices from the start, we need to consider reproducibility methods,
collaborative approaches and transparent communication processes for the
research team as well as the external stakeholders. As project leads,
managers and team organisers, it is crucial to be deliberate and clear
about the tools and platforms selected for the project, as well as
expectations from each contributor from the beginning. Dedicating some
time in thinking through and documenting the setup of a project saves
time, ensuring successful implementation of research plans at different
stages of research. At this stage, you canât be sure that everything
will always go as planned or there will be no unexpected challenges, but
it helps prepare in advance for risk management and adapt to changes
when needed.</p>
</section><section id="main-concerns-and-challenges"><h2 class="section-heading">Main Concerns and Challenges<a class="anchor" aria-label="anchor" href="#main-concerns-and-challenges"></a>
</h2>
<hr class="half-width">
<p>Scientific results and evidence are strengthened if those results can
be replicated and confirmed by several independent researchers. This
means understanding and documenting the research process, describing
what steps are involved, what decisions are made from design to analysis
to implementation stages and publishing them for others to validate.
Research projects already start with multiple documents such as project
proposal, institutional policies and recommendations (including project
timeline, data management plan, open access policy, grant requirements
and ethical committee recommendations), which should be available to the
entire research team at all times. Furthermore, throughout the lifecycle
of a project we handle experimental materials such as data and code,
refer to different published studies, establish collaboration with
others, generate research outputs including figures, graphs and
publications, many of which undergo multiple versions. Then there is a
general need to document the teamâs way of working, different roles and
contribution types, project workflows, research process, learning
resources and templates (such as for presentation, documentation,
project reporting and manuscript) for your research team.</p>
<p>If not planned in advance, these different kinds of information
related to the project can become challenging to record, manage or
retrieve â costing precious time of everyone involved and negatively
affecting collaborative work in your research team.</p>
</section><section id="shared-repository-to-share-information"><h2 class="section-heading">Shared Repository to Share Information<a class="anchor" aria-label="anchor" href="#shared-repository-to-share-information"></a>
</h2>
<hr class="half-width">
<p>To manage collaborative research in computational projects with
mainly distributed systems (different computers, cloud infrastructure,
remote team members) it is essential to provide clear guidelines on
where these digital objects should be held, handled and shared.
Therefore, the first step is to establish a shared digital location
(centralised, findable and accessible) like a shared drive (cloud-based
or organisation-hosted server space) or online repository where all
project related documentation and resources can be made available for
everyone in your research team.</p>
<p>When introduced with clear guidance for how everyone in your team can
contribute to keeping the shared repository up-to-date, it helps build a
sense of collaboration from the start. You can use this repository also
to communicate what policies are relevant for people and their work in
the project; how data, code and documentation are organised; and how
peer-review, open feedback and co-creation will be enabled at all stages
of the project.</p>
<div id="callout1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>The use of standard folder structure and reamde files can help to
provide a similar structure to all project and make sure all information
is gathered and available. One such standard has been developed
following researcherâs feedback inside the <code>gin-tonic</code>
project: <a href="https://gin-tonic.netlify.app/standard/" class="external-link uri">https://gin-tonic.netlify.app/standard/</a></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="versioning">Versioning<a class="anchor" aria-label="anchor" href="#versioning"></a>
</h3>
<p>No matter how your group is organized, the work of many contributors
needs to be managed into a single set of shared working documents.
Version control is an approach to record changes made in a file or set
of files over time so that you and your collaborators can track their
history, review any changes, and revert or go back to earlier versions.
Management of changes or revisions to any types of information made in a
file or project is called versioning.</p>
<p>See the version control episode for detailed information.</p>
</div>
<div class="section level3">
<h3 id="vision-mission-and-milestones">Vision, Mission and Milestones<a class="anchor" aria-label="anchor" href="#vision-mission-and-milestones"></a>
</h3>
<p>It is particularly important to share the projectâs vision, mission
and milestones transparently. Provide sufficient information for what
the expected outcomes and deliverables are. Provide overarching as well
as short-term goals and describe expected outcomes to help contributors
move away from focusing on a single idea of the feature. Describe the
possible expansion of the project to give an idea of what to expect
beyond the initial implementation. All proposed plans for the project
with information on available resources and recommended practices to
ensure everyone is on the same page.</p>
</div>
<div class="section level3">
<h3 id="role-and-responsibilities">Role and Responsibilities<a class="anchor" aria-label="anchor" href="#role-and-responsibilities"></a>
</h3>
<p>Create a folder/directory to give information about the different
team members with their roles in the project, key skills, interests and
contact information (when possible). Describe what responsibilities and
opportunities for collaboration different members will have. Provide
resources on ways of working to ensure fair participation of
stakeholders who collaborate on short- and long-term milestones within
the project. It reduces or addresses concerns about the projectâs
progress towards meeting goals and prevent potential fallout between
project stakeholders. When possible, such as in an open source project,
provide these details for those outside the current group, especially
when you want to encourage people outside the project to be
involved.</p>
<p>You may use a spreadsheet to keep an overview of the members
responsability and contribution to the project. You may want to use the
CrediT taxonomy in that spreadseet, so that this information can be
easily added to the final manuscript and other research outputs
metadata.</p>
</div>
<div class="section level3">
<h3 id="data-management-plan">Data management plan<a class="anchor" aria-label="anchor" href="#data-management-plan"></a>
</h3>
<p>This will be discussed in detail in the Research Data Management
section.</p>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>What research output are you aiming for? Will you publish the data,
the code, the hardware documentation, the protocols independently of the
main manuscript publication?</p>
</div>
</div>
</div>
<blockquote>
<h2 id="start-with-an-intention-to-releasepublish">Start with an
intention to Release/publish</h2>
<ul>
<li>
<strong>Structure and logically organise project folders and
files</strong> using the consistent convention for individual file
names, making them easy to locate, access and reuse.</li>
<li>Review and <strong>consider how research needs to be
disseminated</strong> at the end of the project as per the grant as well
as institution requirements and policies. This include the choice of
licenses for dissemination.</li>
<li>Discuss team member <strong>responsability and roles</strong>, as
well as author byline at the start of the project</li>
<li>
<strong>Embed computational reproducibility</strong>, notably
through skill-building in your team (see version control, computational
environments, code testing, software package management).</li>
<li>
<strong>Add documentation process to project timelines and
milestones</strong> for capturing progress, blockers and contributions
by all stakeholders, making your research objects easy to attribute and
release.</li>
</ul>
</blockquote>
</div>
<div class="section level3">
<h3 id="team-framework">Team Framework<a class="anchor" aria-label="anchor" href="#team-framework"></a>
</h3>
<p>To ensure that all team members have a shared understanding of ways
of working, select or adapt a <a href="https://the-turing-way.netlify.app/collaboration/new-community/new-community-teamwork.html" class="external-link">Team
Framework</a> that provides guidance on how to best work in your team.
For instance, Agile workflow for teamwork enables iterative development,
with frequent interaction between interested parties to decide and
update requirements. See <a href="https://nlesc.github.io/teamwork-for-research-software-development/" class="external-link">Teamwork
for Research Software Development</a> tutorial by Netherlands eScience
Center with lessons on teamwork, agile and scrum framework, project
board such as kanban, challenges and practical recommendations. See next
lessons for more information about team work and collaborative
working.</p>
</div>
</section><section id="a-good-start"><h2 class="section-heading">A good start<a class="anchor" aria-label="anchor" href="#a-good-start"></a>
</h2>
<hr class="half-width">
<p>In addition to ensuring effective development and collaboration
during the lifetime of the project, a well-organised project also
ensures sustainability and reusability of research for both the
developers and future users more dynamically. But it all comes to
implement the tools planned in the setting up phase. One may need to
iteratively modify the plan and include different tools if the team
changes or grow. This aspect is discussed in detail in the Research Data
Management episode.</p>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Create a project on GitHub</li>
</ol>
<ul>
<li>Create a Github account at <a href="https://github.com" class="external-link uri">https://github.com</a>
</li>
<li>Go to <a href="https://github.com/tonic-team/Tonic-Research-Project-Template/" class="external-link uri">https://github.com/tonic-team/Tonic-Research-Project-Template/</a>
</li>
<li>Click <code>Use this template</code> and follow instructions.</li>
</ul>
<p>NB: You can use that template on other git platforms.</p>
<ol start="2" style="list-style-type: decimal">
<li><p>Change the <code>reamde.md</code> file directly on the browser:
enter information about the project, commit with a message. Enter
information about the team (roles and responsabilities), commit with a
second message.</p></li>
<li><p>Look at the history of the readme.md file, understand the need
for good commit messages.</p></li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Shared repository with well structured and organised files are
crucial for starting a project</li>
<li>Documentation is as important as data and code to understand the
different aspects of the project and communicate about the
research.</li>
<li>Licencing and open science practices allow proper use and reuse of
all research objects, hence should be applied in computational research
from the start.</li>
</ul>
</div>
</div>
</div>
<p><br><br><br><br></p>
<hr>
<p><br><br></p>

</section><section id="project-management-tools"><h2 class="section-heading">Project Management Tools<a class="anchor" aria-label="anchor" href="#project-management-tools"></a>
</h2>
<hr class="half-width">
<p>In the previous chapters, we have already discussed practices that
enable the effective management of projects in:</p>
<ul>
<li>setting up shared resources;</li>
<li>defining the vision, mission and roadmap of your project;</li>
<li>managing data and other research-related resources; and</li>
<li>versioning and tracking progress.</li>
</ul>
<p>It is important to communicate tasks and responsibilities to
different stakeholders of the project. However, what is even more
important is to allow all members to understand where in the entire
project their tasks fit and how they can track the progress of the
entire project.</p>
<div class="section level3">
<h3 id="gantt-charts">Gantt charts<a class="anchor" aria-label="anchor" href="#gantt-charts"></a>
</h3>
<p>Gantt charts are timeline views of the project plan. It defines which
task should be done (and finished) at which time (and by whom). Some
advanced tools allows to visualise dependencies between tasks. It is an
important tool to stay on track or redefine priorities when milestones
are not met in time.</p>
</div>
<div class="section level3">
<h3 id="kanban">Kanban<a class="anchor" aria-label="anchor" href="#kanban"></a>
</h3>
<p>Kanban provide a visual overview of the tasks, their status (to do,
in progress, done) and the people responsible for them in a columnar
form. While physical boards and post-it (of different colors and forms)
are often used, these tasks can be visualised on a digital board where
different columns can present different statuses, different task groups
or priorities.</p>
<p>It is also a tool to focus the work on a few tasks. Indeed the âin
progressâ tab should only contain 1-3 elements per team member, so that
one finish one task before starting a new one. Kanban can also be used
to communicate what is in progress to other team members and coordinate
the work.</p>
<p>Some tools that are popular among research community is <a href="https://asana.com/" class="external-link">Asana</a>, <a href="https://trello.com/en-GB" class="external-link">Trello</a>, <a href="https://blog.doist.com/todoist-project-management/" class="external-link">Todoist</a>
and <a href="https://www.notion.so/" class="external-link">Notion</a>.</p>
<p>The majorities of Git repositories have advanced features like
embedded kanban systems, allowing some automation of cards movement and
links between task management and data storage.</p>
</div>
<div class="section level3">
<h3 id="git-repositories-for-project-management">Git repositories for Project Management<a class="anchor" aria-label="anchor" href="#git-repositories-for-project-management"></a>
</h3>
<p>Similar features on GitHub, GitLab, Codeberg or GIN can be used for
project management. The following section takes GitHub as an example,
but most of this can be applied in other platforms.</p>
<p><strong>Issue</strong> is a GitHub integrated feature that allows
everyone to track the progress on GitHub. Similar to a âTo-Do Listâ,
issues can be anything from a project milestone (releasing an R package,
submitting to an online data repository, a working simulation) but also
specific issues with code (fixing a bug, adding a function, updating
tests).</p>
<p>Based on the tasks described in an issue, your collaborators can
address them and save or âcommitâ changes in their local copy of the
repository. Local changes then can be âpushedâ to the repository on
GitHub for âreviewâ via the <strong>Pull Request</strong> feature. Once
a pull request is opened, different collaborators can discuss and review
the potential changes and add follow-up commits before those changes are
âmergedâ into the main repository.</p>
<p><strong>Project boards</strong> are kanban-like features on GitHub
that help you visualise (list of tasks), categorise (in columns) and
prioritise (drag/move around) different tasks. A collection of project
boards can be created for a different set of tasks, comprehensive
roadmaps, or even release checklists. By linking <strong>issues and Pull
Requests</strong>, project boards can create workflows. The Project
board shows metadata for issues and pull requests, like labels,
assignees, the status, and who opened it. Additional notes within
columns can be added as task reminders, references to issues and pull
requests from any repository on GitHub.com, or to add information
related to the project board. This Kanban board feature can be very
helpful in getting a snapshot of multiple research projects within a
team/lab and tracking what multiple people are currently working on. You
can read more about <a href="https://docs.github.com/en/issues/organizing-your-work-with-project-boards/managing-project-boards/about-project-boards" class="external-link">Project
Board in GitHub Documentation</a>.</p>
<figure><img src="https://i.postimg.cc/3wDm4Qw4/Screenshot-2022-02-10-at-19-47-35.png" alt="drawing" width="1000" class="figure mx-auto d-block"></figure><p><em>An example is Kanban for researcher project management. GitHub
boards can be given any name.</em></p>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Tutorial: Kanban Boards for Project Management (Click to view)
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Within Github, the Projects Tab can provide a board with cards to
organise issues collaboratively. If a team is already working within
Github, this can be beneficial as everything remains in the same place.
âIssuesâ can be used as a record of ToDos or others flagging up bugs and
features that need to be addressed. They can be attached to particular
repos and assigned to people.</p>
<p><img src="https://i.postimg.cc/Kzfz40t3/Screenshot-2022-02-10-at-19-55-23.png" alt="drawing" width="600" class="figure"><em>A traditional Kanban for a collaborative computational project.
Keeping track of bugs and what everyone is working on.</em></p>
<p>The Kanban board can be modified to whatever layout or structure
makes sense to you. This example uses the concept to publication/release
pipeline.</p>
<figure><img src="https://i.postimg.cc/3wDm4Qw4/Screenshot-2022-02-10-at-19-47-35.png" alt="drawing" width="1000" class="figure mx-auto d-block"></figure><p>Github also allows different summary views for collaborative issues
across multiple repos, which can be helpful for organising larger
teams.</p>
<figure><img src="https://i.postimg.cc/d37Yv66Y/Screenshot-2022-02-10-at-19-47-51.png" alt="drawing" width="600" class="figure mx-auto d-block"></figure><p><em>Author: Lydia France (Junior Data Scientist, The Alan Turing
Institute, UK)</em></p>
<p>Similar walkthrough can be done with <a href="https://next.forgejo.org" class="external-link uri">https://next.forgejo.org</a>
(Codeberg-like test platform). Lecturer prepare a repository with a
readme, a project and one issue. Fake user should be added to the
platform before the course, so that people do not have to register, and
users are added as collaborators on the repository.</p>
<ul>
<li>create an issue âadding myself to the participant listâ</li>
<li>assign it to the project</li>
<li>move it in the board view</li>
<li>assign someone, add a label, add a milestone.</li>
<li>modify readme file, commit change with a new branch.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
</section><section id="collaborating-on-computational-projects"><h2 class="section-heading">Collaborating on Computational Projects<a class="anchor" aria-label="anchor" href="#collaborating-on-computational-projects"></a>
</h2>
<hr class="half-width">
<p>Much research is now collaborative and a shared code repository can
be effectively used to enable collaboration at all stages of code
development at the analysis and implementation stage.</p>
<p>Later in this material, we will go through best practices in code
writing, testing, reviews and modularity, which help achieving
computational reproducibility. Before that, we will look at ways to
foster documentation of the work, and how research data management is
necessary for efficient collaboration and reproducible research.</p>
</section><section id="document-collaborations"><h2 class="section-heading">Document collaborations<a class="anchor" aria-label="anchor" href="#document-collaborations"></a>
</h2>
<hr class="half-width">
<p>It is important to document who does what in a project, such that
credit can be fairly attributed. This starts at day one, when roles and
responsibilities are set, but needs to be often updated as the roles may
change during the project. In large team, this may be an essential part
of growing a community, especially if one does recognize roles usually
overseen in the research environment.</p>
<p>The tenzing spreadsheet (<a href="https://tenzing.club" class="external-link uri">https://tenzing.club</a>) will allow to present the
contribution in publications and may be a good tool to start the
collection of contribution. A lot of open source project, like The
turing way, are using a bot (<a href="https://allcontributors.org" class="external-link uri">https://allcontributors.org</a>, works only on GitHub) to
collect and show several contribution types.</p>
<div id="discussion3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion3"></a>
</h3>
<div class="callout-content">
<ul>
<li>Go to your research repository created in the setup chapter</li>
<li>Open one issue, assign one person and a milestone. For instance
<code>write a data management plan</code> milestone in 6 months.</li>
<li>add to do to the issue (optional)</li>
<li>Use the project board to move the issue inside the board, move it to
a different column.</li>
</ul>
</div>
</div>
</div>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ul>
<li>Make group leaders familiar with practices that are crucial for
their teams to develop reproducible code.</li>
<li>Encourage researchers to think about code reproducibility through
quality check, testing, sharing their code as well as a research
environment.</li>
<li>Introduce Continuous Integration for automating the testing
process.</li>
</ul>
</div>
</div>
</div>
<div id="discussion4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion4"></a>
</h3>
<div class="callout-content">
<p>What measures described in this session are meant to make research
more reproducible, what measures are meant to facilitate team work. How
does the different measures interact ?</p>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Making interaction measure is also making the research more
reproducible. Indeed most of the reusability of the data and code can be
best tested when several people interact on a project. Documenting for a
current collaborator helps a lot to make sure documentation will be
enough for new collaboratory coming in the project later on.</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="aio-25_RDM"><p>Content from <a href="25_RDM.html">Managing Data</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/25_RDM.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is considered research data?</li>
<li>How to start building a research data management plan?</li>
<li>What is FAIR principles for data management?</li>
<li>Why care about documentation and metadata standards?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe research data management (RDM)</li>
<li>Explain FAIR principles and practices for RDM</li>
<li>Introduce data storage and organisation plan</li>
<li>Discuss documentation and metadata practices</li>
<li>Overview of best practices for spreadsheet design</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="research-data-management"><h2 class="section-heading">Research Data Management<a class="anchor" aria-label="anchor" href="#research-data-management"></a>
</h2>
<hr class="half-width">
<p>Research data management is a huge topic here, because only well
managed data can be well analysed and can be usefully shared with the
community. It raises concerns and challenges, but also opportunities. It
represents about 5% of the time in a research project. Some estimate
that it can save up (on average) about 15% of the research time (less
time lost due to data loss, fewer errors in the data, automated
analysis, data and analysis code reuse).</p>
<div id="invest-in-data-management-because" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="invest-in-data-management-because" class="callout-inner">
<h3 class="callout-title">Invest in data management because:<a class="anchor" aria-label="anchor" href="#invest-in-data-management-because"></a>
</h3>
<div class="callout-content">
<ul>
<li>Managing your data allows you to always find your data and ensure
the quality of scientific practice or research.</li>
<li>Storing your data properly and backing up regularly prevents data
loss.</li>
<li>It can help with recognition for all research outputs.</li>
<li>It stimulates collaboration with others, who will find it easier to
understand and reuse your data.</li>
<li>RDM is cost/time efficient, as you will always be able to find and
use your data.</li>
</ul>
</div>
</div>
</div>
<div id="need-to-consider-standard-file-formats-for-future-use-of-data" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div id="need-to-consider-standard-file-formats-for-future-use-of-data" class="callout-inner">
<h3 class="callout-title">Need to consider standard file formats for
future use of data!<a class="anchor" aria-label="anchor" href="#need-to-consider-standard-file-formats-for-future-use-of-data"></a>
</h3>
<div class="callout-content">
<p><em>Electron Microscope Facility in our institute has produced around
5 petabytes (5,000,000 GB) of data since the institute opening. These
files are stored safely and privately, and have not been standardised.
As a result they are in danger of being lost forever, stored but never
used. With metadata, this could form a transformative training data set
for machine learning tools and possibly lead to new discoveries and
insights. Creating Alpha Fold and other machine learning/AI tools you
need large data sets. Meta data allows data to be future proofed for
further research and even innovative research not currently
possible.</em></p>
</div>
</div>
</div>
</section><section id="overview-of-research-data-management"><h2 class="section-heading">Overview of Research Data Management<a class="anchor" aria-label="anchor" href="#overview-of-research-data-management"></a>
</h2>
<hr class="half-width">
<p>Research Data Management (RDM) covers how research data can be
stored, described and reused. Data here is used as a generic term to
encompass all digital objects. RDM is a vital part of enabling
reproducible research. RDM ensures efficiency in research workflows, and
also greater reach and impact, as data become FAIR (Findable,
Accessible, Interoperable and Reusable).</p>
<p>Data should be stored in multiple locations and backed up regularly
to prevent loss or data corruption. It is also important to consider the
whole life cycle of the data (especially making collected data computer
readable) when designing data management plan and tasks.</p>
<p>Clearly describing data using documentation and metadata ensures that
others know how to access, use and reuse your data, and also enable
conditions for sharing and publishing data to be outlined.</p>
<div class="section level3">
<h3 id="the-research-data-lifecycle---a-model-for-data-management">The Research Data Lifecycle - A Model for Data Management<a class="anchor" aria-label="anchor" href="#the-research-data-lifecycle---a-model-for-data-management"></a>
</h3>
<p>Research data often follows a âlifecycleâ that follows the research
project as it evolves. This model provides a sound basis on which to
plan for research data management, from data creation at the start of a
research project, through to publishing and sharing research at the end
of the project, and archiving any research data for the long-term and
future re-use once the project has ended.</p>
<p>The research data lifecycle involves data creation, data use, data
publication and sharing, data archiving, and data re-use or destruction.
However, data have a longer lifespan than the research project that
creates them. In a Data Management Plan, you can structure how you will
manage and share your research data.</p>
</div>
</section><section id="data-management-plan"><h2 class="section-heading">Data Management Plan<a class="anchor" aria-label="anchor" href="#data-management-plan"></a>
</h2>
<hr class="half-width">
<p>A Data Management Plan (DMP), or Output Management Plan, is a
document that describes how your research outputs will be generated,
stored, used and shared within your project. A DMP is a living document,
which can be updated throughout the research project as needed.</p>
<p>A Data Management Plan is a roadmap for you to manage your data
efficiently and securely. This can prevent data loss or breaches.
Planning ahead on how to manage your data consistently can save you time
later on!</p>
<div class="section level3">
<h3 id="a-data-management-plan-should-provide-information-on-five-main-topics">A Data Management Plan should provide information on five main
topics<a class="anchor" aria-label="anchor" href="#a-data-management-plan-should-provide-information-on-five-main-topics"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Roles and Responsibilities</strong> for the management of
the data and code to help prevent confusion/miscommunication later in
the project. Please check the DMP recommendations and requirements
library research support team of your institute and the website of your
funder. You can check if your funder or institute has a DMP template
using <a href="https://dmponline.dcc.ac.uk/" class="external-link">DMPonline</a>.</li>
<li>
<strong>A list of types, standards and formats</strong> for data,
documentation and metadata (discussed later) should allow team members
to understand to comply with the recommendations from the start of the
project.</li>
</ol>
<ul>
<li>A distinction can be described in the plan separately for different
data types such as raw (primary), processed and ready to use (finalised
to publish) datasets.</li>
<li>All types of data will have to be described to be placed into
context by using metadata and adequate documentation which will allow
anyone in your team to interpret the data in the future.</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Data storage and backup procedures</strong> should be
assessed for each project and established depending on the institutional
requirements, associated costs and recommended format from your field.
We will diuscuss this in detail later in this lesson.</li>
<li>
<strong>Preservation of the research outputs</strong> can be managed
differently based on if they can be made publicly available or not.
Personal data or research outputs needed to apply for patents cannot be
publicly shared but they still have to be preserved for several years,
depending on the policies of your country, institute and funder. Learn
more about this in detail in <a href="https://the-turing-way.netlify.app/reproducible-research/rdm/rdm-sharing.html#rr-rdm-sharing" class="external-link">Sharing
and Archiving Data</a> chapter in <em>The Turing Way</em>.</li>
<li>
<strong>Reuse of your research outputs</strong> should be ensured by
selecting licenses for different components of your research when you
make your output available on a repository (see the Licensing
subchapters on <a href="https://the-turing-way.netlify.app/reproducible-research/licensing/licensing-data.html#rr-licensing-data" class="external-link">data</a>
and <a href="https://the-turing-way.netlify.app/reproducible-research/licensing/licensing-software.html#rr-licensing-software" class="external-link">software</a>
for more information). A dedicated document (such as a README file) is
recommended for describing research outputs into context. UK Data
Service has provided a <a href="https://ukdataservice.ac.uk/learning-hub/research-data-management/plan-to-share/checklist/" class="external-link">Data
Management Checklist</a> to help cover different aspects of the
DMP.</li>
</ol>
<div id="plan-for-reproducible-research" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="plan-for-reproducible-research" class="callout-inner">
<h3 class="callout-title">Plan for reproducible research<a class="anchor" aria-label="anchor" href="#plan-for-reproducible-research"></a>
</h3>
<div class="callout-content">
<p>When planning a reproducible analysis, your data management plan will
change. Because you want to avoid any manual wrangling of the data
(copy-paste, analysis in excel, â¦), you need to take particular care of
the format of the raw data, making it if possible computer readable with
the programming language you will be using. Keep the raw data as raw as
possible so that errors can be avoided and calculations can be changed.
Also when a software provide you with both raw data and pre-analysed
data, you should consider early which file you will be using in your
analysis (in most cases, using the raw data will make your analysis more
robust and easier to reuse for other researchers). Also information may
be lost when transforming the raw data, so make new versions and keep
the original file safe.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="who-writes-a-data-management-plan">Who writes a Data Management Plan ?<a class="anchor" aria-label="anchor" href="#who-writes-a-data-management-plan"></a>
</h3>
<p>At best, the whole team should participate in the planning. It is
often good to have some specialist who knows what are the backup and
publication pathways at the institution or in the research domain. Only
the data collectors will know details about the data, though. The data
analysts need to be involved to make sure the data is relatively easy to
analyse In addition, the researchers will have to implement the plan, so
they need to be aware and to agree with it.</p>
<div id="callout3" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout3"></a>
</h3>
<div class="callout-content">
<p>When the data is meant to be open, researchers with different
questions and needs may try to analyse the data in unexpected ways.
These researchers will not be there at the Data management plan stage
and the researchers will probably not have included their needs.</p>
<p>A way to make the data the most useful is to take the FAIR principles
into account, keep the raw data raw, and add metadata. A critical
metadata component in this case is a contact person, who may be able to
explain the data and its analysis to new researchers.</p>
</div>
</div>
</div>
</div>
</section><section id="defining-data"><h2 class="section-heading">Defining Data<a class="anchor" aria-label="anchor" href="#defining-data"></a>
</h2>
<hr class="half-width">
<p>Data are objects that you use and produce during your research life
cycle, encompassing data sets, software, code, workflow, models,
figures, tables, images and videos, interviews, articles. Data are your
research asset. A good way of thinking about what might be classed as
data that needs to be managed is to ask yourself the questions:</p>
<ul>
<li>What is the information that I need to use and write about in my
paper or book?</li>
<li>What information would I need to back up my conclusions?</li>
<li>What information is needed by others to understand and possibly
replicate the research that I have done? This information is your
data.</li>
</ul>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<ul>
<li>What data are you collecting ?</li>
<li>Is it all digitalised?</li>
<li>Do you work with an open format, which will be readable in 10 years
?</li>
</ul>
</div>
</div>
</div>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>When collecting the Body-mass index (BMI) of participants, how would
you do this ?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Best to record the height and weight of the subject. These are more
raw forms of the data.</p>
<ul>
<li>This is the information the recording devices will give so errors
are avoided.</li>
<li>the BMI may not be the best measure in the future, weight + height
is more informative than the BMI.</li>
</ul>
</div>
</div>
</div>
</div>
</section><section id="using-spreadsheets"><h2 class="section-heading">Using spreadsheets<a class="anchor" aria-label="anchor" href="#using-spreadsheets"></a>
</h2>
<hr class="half-width">
<p>Here we made a special section on spreadsheets design and management,
because <strong>most data analysis nightmares (and time waste) are due
to bad spreadsheet design. </strong></p>
<p>More information is available in the <a href="spreadsheets.html">extra content of this course</a>. Please refer
to the <a href="https://datacarpentry.org/spreadsheet-ecology-lesson/" class="external-link">Data
Carpentry Ecology Lesson</a> and <em>The Turing Way</em> chapter for <a href="https://the-turing-way.netlify.app/reproducible-research/rdm/rdm-spreadsheets.html" class="external-link">managing
data in spreadsheet</a> for best practices.</p>
<p>Spreadsheets, such as Microsoft Excel files, google sheets, and their
Open Source alternative <a href="https://www.libreoffice.org" class="external-link">(for
instance) LibreOffice</a>, are commonly used by wet-lab experimentalists
to collect, store, manipulate, analyse, and share research data.
Spreadsheets are convenient and easy-to-use tools for organising
information into an easy to write and easy to read forms for humans.
However, one should use them with caution, as the use of an
inappropriate spreadsheet is a major cause of mistakes in the data
analysis workflow.</p>
<p>As a project manager, it is important to see that spreadsheet are
only used to collect information, but that manipulation and analysis are
made in a programming environment. This also means that the information
should be understandable by humans, and readable by machines.</p>
<p>This includes:</p>
<ul>
<li>Only one table per spreadsheets</li>
<li>Spreadsheet are stored as .csv or .tsv files (no color, no merging
cells)</li>
<li>Spreadsheet can be read by the programming language in use (they are
tidy)</li>
<li>At best, there is an automatic validation tool is run on the
spreadsheet.</li>
<li>New data should lead to new rows (and not new columns).</li>
</ul>
<figure><img src="fig//tidy-1.png" alt="An illustration of tidy data with three panels with three titles. Arrows are indicating the direction: vertical arrows for variables, as they should be indicated in columns, and horizontal arrows for observations , as they should be indicated in rows. In the third pannel, a circle is present on each cell to represent that there is one unique value per cell." class="figure mx-auto d-block"><figcaption>An illustration of tidy data: variables are in
columns, observations in rows and there is one value per cell.</figcaption></figure></section><section id="file-names-and-folder-structure"><h2 class="section-heading">File names and folder structure<a class="anchor" aria-label="anchor" href="#file-names-and-folder-structure"></a>
</h2>
<hr class="half-width">
<p>While some structure may be given during the project setup, one
should keep track of the sub-structure added, make sure the files are
saved in the correct folder. In addition, one should make sure that file
names are chosen carefully. Indeed, the file name is the first encounter
with a file and making names informative can save a lot of time
especially when searching for a specific file. File names should be both
human and computer readable (conveying information about the file
content, in a way a computer can extract it.)</p>
<p>It is important to have written conventions that are easy to find. <a href="https://the-turing-way.netlify.app/reproducible-research/rdm/rdm-storage.html?highlight=name#file-naming-conventions" class="external-link">turingway
chapter</a></p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>What date is this date : 01.03.1931</p>
<p>January 3st 1931 ? (USA) March 1st 1931 ? (the rest of the world)</p>
<p>How to write the date in an excel spreadsheet ?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>While the standard is 1931-03-01 (YYYY-MM-DD), excel will transform
this into a number, the only safe way to deal with dates in Excel is to
have 3 columns.</p>
</div>
</div>
</div>
</div>
<div id="callout4" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout4"></a>
</h3>
<div class="callout-content">
<p>Naming files tips:</p>
<ul>
<li>names should be unique in a project</li>
<li>file location should not be needed to understand what the file is
about</li>
<li>as short as possible, as long as necessary</li>
<li>avoid abbreviation if possible, document them well otherwise.</li>
<li>when using dates, use ISO format: yyyy-mm-dd (yyyymmdd can work
too)</li>
<li>no special characters (no space) in filenames, use - or _</li>
<li>use only lowercase or be very careful (linux versus other
systems)</li>
</ul>
<p>see <a href="https://speakerdeck.com/jennybc/how-to-name-files" class="external-link uri">https://speakerdeck.com/jennybc/how-to-name-files</a></p>
</div>
</div>
</div>
</section><section id="documentation-and-metadata"><h2 class="section-heading">Documentation and Metadata<a class="anchor" aria-label="anchor" href="#documentation-and-metadata"></a>
</h2>
<hr class="half-width">
<p>Having data available is of no use if it cannot be understood.
Therefore research data should always include consistent documents and
metadata.</p>
<p>Data documentation provides contexts and full description about the
data. It allows your collaborators, colleagues and future you to
understand what has been done and why. Ideally written in clear and
plain language, documentation describes data with sufficient information
such as source, strengths, weaknesses, and analytical limitations of the
data allowing users to make informed decisions when using it.</p>
<p>Without metadata to provide provenance and context, the data canât be
used effectively. Metadata is information about the data, descriptors
that facilitate cataloguing data and data discovery. Often, metadata are
intended for machine reading. When data is submitted to a trusted data
repository, the machine-readable metadata is generated by the
repository. If the data is not in a repository a text file with
machine-readable metadata can be added as part of the documentation.</p>
<ul>
<li>The type of research and the nature of the data also influence what
kind of documentation is necessary. It is best practice to use
recognised community metadata standards to make it easier for datasets
to be combined.</li>
<li>The level of documentation and metadata will vary according to the
project type and the range of stakeholders, including the future
users.</li>
<li>Examples of documentation may include items like <a href="https://help.osf.io/hc/en-us/articles/360019739054-How-to-Make-a-Data-Dictionary" class="external-link">data
dictionaries</a> or codebooks, protocols, logbooks or lab journals,
README files, research logs, analysis syntax, algorithms and code
comments.</li>
<li>Variables should be defined and explained using data dictionaries or
codebooks.</li>
<li>Data should be stored in logical and hierarchical folder structures,
with a README file used to describe the structure.</li>
</ul>
<div id="rembi-example-of-metadata-in-bioimaging-data" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div id="rembi-example-of-metadata-in-bioimaging-data" class="callout-inner">
<h3 class="callout-title">REMBI: Example of metadata in bioimaging
data<a class="anchor" aria-label="anchor" href="#rembi-example-of-metadata-in-bioimaging-data"></a>
</h3>
<div class="callout-content">
<p>REMBI: <strong>Recommended Metadata for Biological Imagesâenabling
reuse of microscopy data in biology</strong>, Sarkans, U., Chiu, W.,
Collinson, L., Darrow, M. C., Ellenberg, J., Grunwald, D., â¦Brazma, A.
(2021). Nature Methods, 18(12), 1418â1422. doi: <a href="https://www.nature.com/articles/s41592-021-01166-8" class="external-link">10.1038/s41592-021-01166-8</a></p>
<p>Bioimaging data have significant potential for reuse, but unlocking
this potential requires systematic archiving of data and metadata in
public databases. Cryo-EM and cryo-ET have proven to be powerful tools
for determining high-resolution structures of biological matter and
examining the functional cellular context of macromolecular
complexes.</p>
<p>REMBI is a draft metadata guidelines to begin addressing the needs of
diverse communities within light and electron microscopy. The current
version of REMBI, including examples from the fields covered by the
three working groups is shared online via <a href="https://docs.google.com/spreadsheets/d/1Ck1NeLp-ZN4eMGdNYo2nV6KLEdSfN6oQBKnnWU6Npeo/edit#gid=1023506919" class="external-link">http://bit.ly/rembi_v1</a>.</p>
</div>
</div>
</div>
</section><section id="data-storage-organisation-and-backup-procedures"><h2 class="section-heading">Data Storage, Organisation and Backup Procedures<a class="anchor" aria-label="anchor" href="#data-storage-organisation-and-backup-procedures"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/backup.jpg" alt='Illustration of backup need with the legend "a good advice". One big character is crying "always make a backup" loud. A second character is startled and let their computer labelled "all my work" fall on the floor and crash.' class="figure mx-auto d-block"><figcaption>BAckup, by Auke Herrema, CC-BY</figcaption></figure><p>Data loss can be catastrophic for your research project and can
happen often. You can prevent data loss by picking suitable storage
solutions and backing your data up frequently.</p>
<ul>
<li>Most institutions will provide a network drive that you can use to
store data.</li>
<li>Portable storage media such as memory sticks (USB sticks) are more
risky and vulnerable to loss and damage.</li>
<li>Cloud storage provides a convenient way to store, backup and
retrieve data. You should check terms of use before using them for your
research data.</li>
</ul>
<p>Especially if you are handling personal or sensitive data, you need
to ensure the cloud option is compliant with any data protection rules
the data is bound by. To add an extra layer of security, you should
encrypt devices and files where needed. Your institution might provide
local storage solutions and policies or guidelines restricting what you
can use. Thus, we recommend you familiarise yourself with your local
policies and recommendations.</p>
<div id="note" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note<a class="anchor" aria-label="anchor" href="#note"></a>
</h3>
<div class="callout-content">
<ul>
<li>Some concepts discussed in the previous chapter such as setting up
project repository, version controlling, pre-registration, and licensing
apply to this point.</li>
<li>Also consider FAIR practices, data organisation and handling
sensitive data practices, as well as metadata and documentation that are
discussed below.</li>
</ul>
</div>
</div>
</div>
<div class="section level3">
<h3 id="backups">Backups<a class="anchor" aria-label="anchor" href="#backups"></a>
</h3>
<p>To avoid losing your data, you should follow good backup
practices.</p>
<ul>
<li>You should have 2 or 3 copies of your files, stored on</li>
<li>at least 2 different storage media,</li>
<li>in different locations.</li>
</ul>
<p>The more important the data and the more often the datasets change,
the more frequently you should back them up. If your files take up a
large amount of space and backing up all of them proves to be
challenging or expensive, you may want to create a set of criteria for
when you back up the data. This can be part of your data management plan
(DMP).</p>
<p>When you are ready to release the data to the wider community, you
can also search for the appropriate databases and repositories in <a href="https://fairsharing.org/databases" class="external-link">FAIRsharing</a>, according to
your data type, and type of access to the data. Learn more about this in
<em>The Turing Way</em> chapter on <a href="https://the-turing-way.netlify.app/reproducible-research/rdm/rdm-sharing.html#rr-rdm-sharing" class="external-link">sharing
and archiving Data</a>.</p>
<!--
[Add recommendations from 1:1 interviews, open communities such as Open Life Science and The Turing Way.]-->
</div>
</section><section id="the-fair-principles"><h2 class="section-heading">The FAIR Principles<a class="anchor" aria-label="anchor" href="#the-fair-principles"></a>
</h2>
<hr class="half-width">
<p>The FAIR guiding principles for scientific data management and
stewardship are guidelines to improve the <strong>F</strong>indability,
<strong>A</strong>ccessibility, <strong>I</strong>nteroperability and
<strong>R</strong>eusability of digital assets; all of which support
research reproducibility. The FAIR principles facilitate the
availability of research data so that others can reuse data.</p>
<div id="fair-data-should-be" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="fair-data-should-be" class="callout-inner">
<h3 class="callout-title">FAIR data should be:<a class="anchor" aria-label="anchor" href="#fair-data-should-be"></a>
</h3>
<div class="callout-content">
<ul>
<li>
<strong>Findable</strong>: The first step in (re)using data is to
find them! Descriptive metadata (information about the data such as
keywords) are essential.</li>
<li>
<strong>Accessible</strong>: Once the user finds the data and
software they need to know how to access it. Data could be openly
available but it is also possible that authentication and authorisation
procedures are necessary.</li>
<li>
<strong>Interoperable</strong>: Data needs to be integrated with
other data and interoperate with applications or workflows.</li>
<li>
<strong>Reusable</strong>: Data should be well-described so that
they can be used, combined, and extended in different settings.</li>
</ul>
<figure><img src="fig/fair-principles.jpg" alt="Four FAIR principles depicted." class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<p>Making data âFAIRâ is not the same as making it âopenâ. Accessible
means that there is a procedure in place to access the data. Data should
be as open as possible, and as closed as necessary. It is also important
to say that the FAIR principles are aspirational: they do not strictly
define how to achieve a state of FAIRness, but rather describe a
continuum of features, attributes, and behaviours that will move a
digital resource closer to that goal. Even though the FAIR principles
have been defined to allow machines to find and use digital objects
automatically, they improve the reusability of data by humans as well.
The capacity of computational systems to find, access, interoperate, and
reuse data, with minimal human intervention, is essential in todayâs
data-driven era.</p>
<p>You can find a more detailed <a href="https://www.go-fair.org/fair-principles" class="external-link">overview of the FAIR
principles by GO FAIR</a> of what the FAIR principles recommend.</p>
<div class="section level3">
<h3 id="summary-of-fair---how-to">Summary of âFAIR - How Toâ<a class="anchor" aria-label="anchor" href="#summary-of-fair---how-to"></a>
</h3>
<blockquote>
<p>We have provided an additional lesson to discuss the How-Tos of FAIR
principles in the context of data and software. See <a href="04-FAIRHowTo.html">FAIR How-To for data and software</a> for
detail.</p>
<ul>
<li>Reference: E. L.-Gebali, S. (2022).
BOSSConf_2022_Research_Data_Management. Zenodo. doi: <a href="https://doi.org/10.5281/zenodo.6490583" class="external-link">10.5281/zenodo.6490583</a>
</li>
</ul>
</blockquote>
<ul>
<li>Deposit your data where others can find it, keep in mind where your
peers can find it, such as, field specific repository and give it a
stable unique identifier (PID).</li>
<li>Make your data &amp; metadata accessible via standard means such as
http/API.</li>
<li>Create metadata and explain in detail what this data is about, never
assume people know!</li>
<li>Deposit metadata with PID and make it available with/out data
i.e.Â in case data itself is heavily protected.</li>
<li>Include information on ownership, provenance and citation.</li>
<li>Outline what the reusers of your data are/not allowed to do, use
clear license. Commonly used licenses like MIT or Creative Commons (keep
in mind funders requirements).</li>
<li>Specify access conditions, if authentication or authorization is
required.</li>
<li>Describe your data in a standardized fashion using agreed
terminology and vocabulary.</li>
<li>Share the data in preferred &amp; open file formats.</li>
<li>Start the process early on!</li>
</ul>
</div>
</section><section id="a-note-on-personal-or-sensitive-data"><h2 class="section-heading">A Note on Personal or Sensitive Data<a class="anchor" aria-label="anchor" href="#a-note-on-personal-or-sensitive-data"></a>
</h2>
<hr class="half-width">
<p>Personal data is information about <strong>living people</strong> who
can be identified using the data that you are processing, either
directly or indirectly. A personâs name, address or Social Security
number, as well as, racial/ethnic identity, political opinions,
religious/philosophical beliefs, trade union membership, genetic and
biometric data, physical or mental health, and sexual orientation are
some examples of personal data. Indirect identifiers include health,
economic, cultural or social characteristics. Especially when a certain
combination of these identifiers can be used to identify a person, care
must be taken to manage the data properly.</p>
<p>There are various policies in place in different countries to protect
the rights of individuals over their personal data. For instance, in the
European Union and the UK the <strong>GDPR</strong> (General Data
Protection Regulation) applies to the processing of personal data and
may require researchers to carry out a Data Protection Impact Assessment
(<a href="https://youtu.be/YRiCb3unz3g?t=988" class="external-link">DPIA</a>). Processing
means doing anything with a personâs information, including collection,
storage, analysis, sharing, deletion and destruction. Please review the
national/institutional policies that apply to your research to ensure
that you are up to date with the requirements of managing sensitive
data. Please read <a href="https://the-turing-way.netlify.app/reproducible-research/rdm/rdm-personal.html" class="external-link">Personal
data management</a>, <a href="https://the-turing-way.netlify.app/reproducible-research/rdm/rdm-personal.html#informed-consent" class="external-link">informed
consent</a>, <a href="https://the-turing-way.netlify.app/ethical-research/ethics-committees.html" class="external-link">Research
Ethics Committees Processes</a> and <a href="https://the-turing-way.netlify.app/reproducible-research/open/open-data.html" class="external-link">Open
Data</a> sections in <em>The Turing Way</em> for further details.</p>
</section><section id="acknowledgement"><h2 class="section-heading">Acknowledgement<a class="anchor" aria-label="anchor" href="#acknowledgement"></a>
</h2>
<hr class="half-width">
<ul>
<li>This episode is a reuse of <em>The Turing Way</em> chapter on <a href="https://the-turing-way.netlify.app/reproducible-research/rdm.html" class="external-link">Research
Data Management</a>.</li>
<li>Among many authors, reviewers and contributors, we thanks Esther
Plomp (Data Steward at TU Delft) for maintaining this chapter in The
Turing Way Guide for Research Data Management.</li>
<li>Please cite <em>The Turing Way</em> as: The Turing Way Community.
(2021). The Turing Way: A handbook for reproducible, ethical and
collaborative research (1.0.1). Zenodo. <a href="https://doi.org/10.5281/zenodo.5671094" class="external-link">https://doi.org/10.5281/zenodo.5671094</a>
</li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Data management plan is produced by the whole team.</li>
<li>All information is present and digitalised</li>
<li>Raw data is raw</li>
<li>Spreadsheets are tidy, validated and in text format (.tsv)</li>
<li>Data is safe (backup)</li>
<li>Data is FAIR (findable, accessible, interoperable and reusable)</li>
<li>Data can be opened and analysed in a programming language</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-26_Coding"><p>Content from <a href="26_Coding.html">Managing code</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/26_Coding.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li><p>What is the role of data wrangling?</p></li>
<li><p>What is literate programming?</p></li>
<li><p>How to use data visualisation for insight and
communication?</p></li>
<li><p>What are the main objectives and best practices for testing and
reviewing code?</p></li>
<li><p>When can continuous integration help?</p></li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the basics of coding</li>
<li>Understand the advantages of modular code</li>
<li>How to foster good practices:
<ul>
<li>using code review</li>
<li>using tests</li>
</ul>
</li>
<li>Discuss the importance of code quality, modular programming, and
code testing for reusable error-free code.</li>
<li>Encourage researchers to combine code with documentation to
communicate their work.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="basics-of-coding"><h2 class="section-heading">Basics of coding<a class="anchor" aria-label="anchor" href="#basics-of-coding"></a>
</h2>
<hr class="half-width">
<p>As a project manager, you do not need to code or know how coding
works. However, it is important to understand the data workflow and
foster documentation, as well as coding best practices, such that you
can understand the analysis.</p>
<p>This chapter gives a small introduction into coding and data analysis
workflow, code modularity, and then how foster good code with tests and
review.</p>
</section><section id="analysis-workflow"><h2 class="section-heading">Analysis workflow<a class="anchor" aria-label="anchor" href="#analysis-workflow"></a>
</h2>
<hr class="half-width">
<p>A usual workflow involves some data processing, before the actual
analysis can be done. Statistical analysis and figure production should
be decided before the data is looked at to prevent harking. Explorative
analysis (looking for correlation and interpretation of the results) can
be done anytime, as long as no hypothesis is rejected from that
analysis.</p>
<figure><img src="fig/dataflow.png" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="data-pre-processing">Data (pre-)processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h3>
<p>Code that cleans and processes data (<strong>processing
code</strong>) provides the very beginning of the data analysis
pipeline: starting with raw data and resulting in processed data. When
the data was not collected in a computer readable format, or when
metadata is missing, this step can be the most time consuming of
all.</p>
<blockquote>
<p>Data wrangling involves cleaning data so it can be easily read and
analysed by machines. It can also involve integration, extraction,
removing missing points, and anything that makes data useable and
functional. Regardless of the methods, the code involved with data
cleaning steps should be carefully documented so that the steps involved
can be repeated from raw data to cleaned data.</p>
</blockquote>

<p>Data wrangling can be a manual or automated process. In scenarios
where datasets are exceptionally large, automated data cleaning becomes
a necessity.</p>
<p>Cleaning data means it can be easily read and analysed by machines
and used in analysis pipelines. It can involve changing labels,
subsetting, integration, extraction, removing missing points, and
anything that makes data usable and functional.</p>
<p>Regardless of the methods, the code involved in data cleaning steps
should be carefully documented so that the steps involved can be
repeated from raw data to clean data. When reviewing this type of code,
consider whether the steps involved are readable and in the correct
order. Especially, when filtering data out, this should be consistent
and indicated in the main method description.</p>
</div>
</section><section id="data-visualisation-analysis-and-statistics"><h2 class="section-heading">Data visualisation, Analysis and Statistics<a class="anchor" aria-label="anchor" href="#data-visualisation-analysis-and-statistics"></a>
</h2>
<hr class="half-width">
<div class="warnings">
<p>Visualising data before designing the data analysis is a form of
harking, as you will cherry pick your data (filtering some data out) or
cherry pick an analysis (what looks promising). You should always design
your analysis before visualising the data, or use visualised data as
training data, not included in the final analysis.</p>
</div>
<p>With readable, clean, processed data, the next stage of the data
pipeline is analysis.</p>
<p>Depending on your computational project, this may involve elaborate
and complex analyses, modelling, simulation, and even machine learning.
However, even if this step is just running a single statistics test,
keeping the code documented and modular in clearly defined steps is
key.</p>
<p>Here is an example of applying a Butterworth filter to some data in
Python. The specifics donât matter, you can consider this code
pseudocode for any kind of analysis step.</p>
<pre><code>genomeProject/analysis/01_butterworth_filter.py

#### 01_butterworth_filter, v1.0
##this code read processed data and apply a low pass filter,
## the output variable is called  filtered_data

#  Import dependencies
import numpy as np
from scipy.signal import butter,filtfilt

# (A) Read processed data from file
df = pd.read_csv('genomeProject/data/220103_GenomicData_processed.csv')

# (B) Defining filter parameters
T = 5.0         # Sample Period
fs = 30.0       # sample rate, Hz
cutoff = 2      # desired cutoff frequency of the filter, Hz
nyq = 0.5 * fs  # Nyquist Frequency
order = 2       # quadratic
n = int(T * fs) # total number of samples

# (C) Define the Butterworth Filter Function
def butter_lowpass_filter(data, cutoff, fs, order):
    normal_cutoff = cutoff / nyq

    # Get the filter coefficients 
    b, a = butter(order, normal_cutoff, btype='low', analog=False)
    filtered_data = filtfilt(b, a, data)

    return filtered_data

# (D) Apply Butterworth Filter Function to the data
filtered_data = butter_lowpass_filter(df, cutoff, fs, order)
</code></pre>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<pre><code>genomeProject/analysis/01_bf.py

import numpy as np
from scipy.signal import butter,filtfilt

df = pd.read_csv('genomeProject/data/220103_GenomicData_processed.csv')

normc = 2 / 15
b, a = butter(2, normc, btype='low', analog=False)
filtered_data = filtfilt(b, a, df)

</code></pre>
<p>The code above will give the same result, why is the first one better
on the long run ?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ul>
<li>easier to read for non-coders</li>
<li>commented</li>
<li>variables easy to change</li>
<li>better names</li>
<li>function reusable + possibility to write tests</li>
</ul>
</div>
</div>
</div>
</div>
<ol style="list-style-type: upper-alpha">
<li>the processed data is read into the script. Next, (B) the fixed
parameters are set and named with comments. These fixed numbers are
saved as variables with names. In (C) the filter itself is written as a
Python Function, which means it can be called multiple times throughout
the script. Because the parameters are not written into the function
directly (so it doesnât say
<code>b, a = butter(2, 15,btype='low', analog=False)</code> but instead
uses variables) this code is reusable without having to paste and edit
the numbers every time you apply a function.</li>
</ol>
<p>You can also call this function in other scripts. It can make sense
to produce a file with the functions inside that can be imported into
different scripts in case other projects also have similar methods. This
is known as a package or library. This means altering a function doesnât
mean searching across every file on every project and changing it dozens
of times.</p>

<p>The output of the analysis code may be statistics results that are
reported in a paper, and therefore the steps required to reproduce them
are critically important.</p>
<div class="section level3">
<h3 id="figures-for-communicating-results">Figures for Communicating Results<a class="anchor" aria-label="anchor" href="#figures-for-communicating-results"></a>
</h3>
<blockquote>
<p>When working with data sets, <code>ggplot</code> (in R) or
<code>matplotlib</code>/<code>seaborn</code> (in Python) libraries
provide attractive figures that can be produced very quickly.
Visualising data should not not wait for the point of publication, and
can be used to explore data from the start, and also illustrate
methodology. This is particularly valuable in Jupyter Notebooks. Code to
produce figures should be literate, functional, reuseable in the same
way as data cleaning and analysis code. That way future visualisations
can be easily updated or reused.</p>
</blockquote>
<p>With the analysis complete, data visualisation is usually used to
communicate results. The code used to produce figures is often the next
step in the data pipeline.</p>
<p>For publications or posters, well-constructed figures improve science
communication and help improve the impact of your research. Being able
to produce multipanel figures with annotations and different colour
schemes is complex but one of the advantages of learning a data science
language.</p>
<p>It is therefore worth taking the time for researchers to learn the
technical skills in R, Python, or another language to produce
visualisations. Producing figures in Excel is limiting and often
frustrating, particularly as there are only limited options in layout
and type of figure.</p>
<figure><img src="https://www.biorxiv.org/content/biorxiv/early/2021/12/19/2021.12.16.473013/F1.large.jpg?width=800&amp;height=600&amp;carousel=1" alt="box plots" width="500" class="figure mx-auto d-block"></figure><blockquote>
<p><em>Exploring constraints on a wetland methane emission ensemble
(WetCHARTs) using GOSAT observations, Parker et al 2020</em></p>
<p><a href="https://doi.org/10.5194/bg-17-5669-2020" class="external-link">https://doi.org/10.5194/bg-17-5669-2020</a></p>
</blockquote>
<figure><img src="https://www.pnas.org/cms/10.1073/pnas.1912459116/asset/86065128-2b04-47e8-ba0f-6a649ff72e3e/assets/graphic/pnas.1912459116fig04.jpeg" alt="box plots" width="600" class="figure mx-auto d-block"></figure><blockquote>
<p><em>Spatial transcriptome profiling by MERFISH reveals subcellular
RNA compartmentalization and cell cycle-dependent gene expression, Xia
et al 2019</em> <em><a href="https://www.pnas.org/doi/10.1073/pnas.1912459116" class="external-link">https://www.pnas.org/doi/10.1073/pnas.1912459116</a></em></p>
</blockquote>
<figure><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41593-019-0559-0/MediaObjects/41593_2019_559_Fig1_HTML.png" alt="box plots" width="600" class="figure mx-auto d-block"></figure><p>These figures are more than just visualising data, theyâre about
communication and require adjusting the styles and formats within
<code>ggplot</code> or <code>matplotlib</code> or other libraries.</p>
<p>As before, any code used to produce visualisations should be
reproducible and literate. Often in peer review figures need to be
adjusted or altered, and having the code to do so makes the process much
simpler.</p>
<p>It is usually cleaner to keep data visualisation code separate from
analysis, just to keep a code base organised and modular.</p>
</div>
<div class="section level3">
<h3 id="accessibility">Accessibility<a class="anchor" aria-label="anchor" href="#accessibility"></a>
</h3>
<ul>
<li>For simple figures, using shaded vs unshaded and single colour is
best when considering publications may be printed in black and
white</li>
<li>Colours should be colourblind friendly, <a href="https://colorbrewer2.org/" class="external-link">resources are available</a> for
this.</li>
</ul>
<figure><img src="https://ristretto.black/wp-content/uploads/2018/12/palette_recommended-1-1024x374.png" alt="box plots" width="500" class="figure mx-auto d-block"></figure><ul>
<li>If using a colour map, avoid the standard rainbow. Other rainbows
like <a href="">viridis colour maps</a> can be downloaded. The
brightness varies across the rainbow which leads to visual artefacts
that do not exist:</li>
</ul>
<figure><img src="https://www.researchgate.net/profile/Mohammad-Raji/publication/311530664/figure/fig1/AS:841506323705857@1577642150559/Typical-example-illustrating-the-aesthetics-of-a-good-b-vs-a-bad-a-colormap-Color.png" alt="box plots" width="400" class="figure mx-auto d-block"></figure><blockquote>
<p><em>The transition between greens to yellow and red are more
prominent than the transitions between greens and blues, making stark
boundaries that do not exist. A rainbow with equal brightness solves the
problem.</em></p>
<p><em>Sisneros et al (2016) Chasing Rainbows: A Color-Theoretic
Framework for Improving and Preserving Bad Colormaps. <a href="https://doi.org/10.1007/978-3-319-50835-1_36" class="external-link">https://doi.org/10.1007/978-3-319-50835-1_36</a></em></p>
</blockquote>
<p>Also one should keep readability in mind with text size/font and
similar considerations</p>
<ul>
<li>Data Visualisation as a tool
<ul>
<li>Reference: <a href="https://helenajambor.wordpress.com/2022/01/04/science-visualization-trends-of-2021/" class="external-link">https://helenajambor.wordpress.com/2022/01/04/science-visualization-trends-of-2021/</a>
</li>
</ul>
</li>
</ul>
</div>
</section><section id="statistical-analysis"><h2 class="section-heading">Statistical analysis<a class="anchor" aria-label="anchor" href="#statistical-analysis"></a>
</h2>
<hr class="half-width">
<p>Selection of appropriate statistical method is very important step in
analysis of data. A wrong selection of the statistical method not only
creates some serious problem during the interpretation of the findings
but also affects the conclusion of the study. In statistics, for each
specific situation, statistical methods are available to analysis and
interpretation of the data. To select the appropriate statistical
method, one need to know the assumption and conditions of the
statistical methods, so that proper statistical method can be selected
for data analysis. Other than knowledge of the statistical methods,
another very important aspect is nature and type of the data collected
and objective of the study because as per objective, corresponding
statistical methods are selected which are suitable on given data.
Incorrect statistical methods can be seen in many conditions like use of
unpaired t-test on paired data or use of parametric test for the data
which does not follow the normal distribution, etc., Two main
statistical methods are used in data analysis: descriptive statistics,
which summarizes data using indexes such as mean, median, standard
deviation and another is inferential statistics, which draws conclusions
from data using statistical tests such as studentâs t-test, ANOVA test,
and many others</p>
<p>Statistics goes outside the scope of this course.</p>
</section><section id="coding-best-practices"><h2 class="section-heading">Coding best practices<a class="anchor" aria-label="anchor" href="#coding-best-practices"></a>
</h2>
<hr class="half-width">
<p>The order of a script is usually:</p>
<ul>
<li>main comments on the script function and goals</li>
<li>importing all dependencies</li>
<li>reading inputs, naming outputs</li>
<li>listing all variables</li>
</ul>
<p>The latter element is particularly important. Variables are often a
decision taken (for instance it could be some threshold value used for
filtering data out, a type of low-pass filter used, or a time window
that sub-set the data). Defining them at the start of the code makes it
easier to change and spot.</p>
<p>In practice, your figure and analysis may well depend on the version
of the data, code and these variables, so it is often interesting to
include this information into the output of the code. This may take the
form of a pdf file, with the figure on the first page, and a table of
variables and metadata on the following pages.</p>
</section><section id="literate-programming"><h2 class="section-heading">Literate Programming<a class="anchor" aria-label="anchor" href="#literate-programming"></a>
</h2>
<hr class="half-width">
<p>Literate programming is about comments and documentation and telling
other humans what is happening in your pipeline. Depending on the scale
of your computational projects, you may use one or multiple of these
options:</p>
<ul>
<li>Inline comments when writing code (directly written in the script
file)</li>
<li>A README file describing what your code does</li>
<li>An online documentation as a user and developer guide with
step-by-step explanation</li>
<li>Quarto/RMarkdown or Jupyter Notebook</li>
</ul>
<p>Most of these files can be written in Markdown.</p>
<p>Often, literate programming refers to these notebooks form, where the
documentation and the code are integrated in one file, which produce one
output (html of pdf most of the time).</p>
<p>This course can be thought as a literate programming content. Its
source is markdown and Rmarkdown files, its output is a pdf.</p>
</section><section id="resources-for-taking-this-to-next-level"><h2 class="section-heading">Resources for taking this to next level<a class="anchor" aria-label="anchor" href="#resources-for-taking-this-to-next-level"></a>
</h2>
<hr class="half-width">
<ul>
<li><a href="https://the-turing-way.netlify.app/collaboration/new-community.html" class="external-link">https://the-turing-way.netlify.app/collaboration/new-community.html</a></li>
<li>Data Visualisation as a tool
<ul>
<li>Reference: <a href="https://helenajambor.wordpress.com/2022/01/04/science-visualization-trends-of-2021/" class="external-link">https://helenajambor.wordpress.com/2022/01/04/science-visualization-trends-of-2021/</a>
</li>
</ul>
</li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Can you follow the analysis workflow without reading the code</li>
<li>Is the code well commented and structures</li>
<li>Are the figure created accessible</li>
<li>Is the statistical analysis founded</li>
</ul>
</div>
</div>
</div>
<p><br><br><br><br></p>
<hr>
<p><br><br></p>
<div class="section level3">
<h3 id="modular-programming-functions">Modular Programming (Functions)<a class="anchor" aria-label="anchor" href="#modular-programming-functions"></a>
</h3>
<!--
- do a diagram of directories
- no copying and pasting
- mindset stable and not changed immutably
-->
<div id="testimonial1" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial1"></a>
</h3>
<div class="callout-content">
<p><em>My postdoc wants to work with messy genomics data. I know my
previous postdoc had to do the same thing and it took her monthsâ¦. but
itâs difficult to read her files so my new postdoc will have to work it
out again.</em></p>
</div>
</div>
</div>
<p>Applying methods from one personâs work and applying it to another
problem can take weeks, if not months, of work. Applying methods from
publications is even harder: static PDF files canât describe the lines
of code and data that lead to those discoveries. This is an increasingly
important problem in the face of growing mistrust in science, and a
reproducibility crisis plaguing the sciences.</p>
<p>Instead, functional programming is about writing code that works as
modular steps. Each step is clearly commented on and carefully produced
so that it can be reused in different contexts. Often when you are
analysing data, you need to repeat the same task many times. For
example, you might have several files that all need loading and clean in
the same way, or you might need to perform the same analysis for
multiple species or parameters. Rather than copying and pasting, writing
a function and calling that function leads to fewer errors and confusion
overall.</p>
<p>We can think of this on a broad scale, say one studentâs
computational work has the following steps, where blue shows data
cleaning, and yellow the analysis and statistics.</p>
<figure><img src="fig/IMP-pipeline1.png" alt="drawing" width="600" class="figure mx-auto d-block"></figure><p>Another student can take reuse the data cleaning and initial
visualisation steps because her data was from the same source and is in
the same format. She can later add her own model:</p>
<figure><img src="fig/IMP-pipeline2.png" alt="drawing" width="600" class="figure mx-auto d-block"></figure><p>On the micro-scale, functional programming ensures that each code
file itself is comprised of modular blocks, whether for data processing,
analysis pipeline, simulation and so on. Depending on your programming
language, these may be used as a package or a library or saved in files
that are available for installation. Just the same as the diagram above,
making sure functions are robust and reuseable means they can be shared
throughout different workflows and for different projects.</p>
<blockquote>
<p>Training in functional programming is usually an excellent
pre-requisite for members of your lab.</p>
</blockquote>
<p>A first step can be to <strong>draw out</strong> and create diagrams
to plan code before starting and identifying the modular steps involved.
This does not require technical knowledge of a language and is,
therefore, a great exercise for direct supervision. You can find
practical details on reproducible code in the <a href="https://www.britishecologicalsociety.org/publications/guides-to" class="external-link">Guides
to Better Science by British Ecological Society</a>.</p>
<p><br><br><br><br></p>
<hr>
<p><br><br></p>
</div>
<div class="section level3">
<h3 id="code-testing">Code Testing<a class="anchor" aria-label="anchor" href="#code-testing"></a>
</h3>
<blockquote>
<p>You should not skip writing tests because you are short on time, you
should write tests because you are short on time.</p>
</blockquote>
<p><img src="https://the-turing-way.netlify.app/_fig/error-management.jpg" class="figure"><em>The Turing Way project illustration by Scriberia. Used under a CC-BY
4.0 licence. DOI: 10.5281/zenodo.3332807.</em></p>
<blockquote>
<p>Writing a test to prove things work right now is useless (duh, you
know what use case you designed it for, of course it will work).
Defending your project against the future idiot who is going to f things
up (most likely yourself) is the right attitude and will lead to proper
tests. Love it! <a href="https://mstdn.social/@flip1/108951972043173555" class="external-link uri">https://mstdn.social/@flip1/108951972043173555</a></p>
</blockquote>
<p>It is very, very easy to make mistakes when coding. A single wrong
use of a character can cause a programâs output to be entirely wrong.
Missing one data point, writing plus instead of minus symbol or using
feet instead of meters might be a genuine human mistake, but in
research, the results can be catastrophic. Careers can be damaged/ended,
vast sums of research funds can be wasted, and valuable time may be lost
exploring incorrect avenues. This is why code testing is vital.</p>
<p>Testing is a learned skill that needs to become a part of working
on/improving a project. After changing their code, researchers should
always check that their changes or fixes have not broken anything. It
also helps deciding when dependencies can be updated. There are several
different kinds of testing and each has best practices specific to
them.</p>
<div id="a-few-important-testing-types" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="a-few-important-testing-types" class="callout-inner">
<h3 class="callout-title">A few important testing types<a class="anchor" aria-label="anchor" href="#a-few-important-testing-types"></a>
</h3>
<div class="callout-content">
<ul>
<li>
<strong>Smoke testing</strong>: Very brief initial checks that
ensure the basic requirements required to run the project hold. If these
fail there is no point in proceeding to additional levels of testing
until they are fixed.</li>
<li>
<strong>Unit testing</strong>: A level of the software testing
process where individual units of a software are tested. The purpose is
to validate that each unit of the software performs as designed.</li>
<li>
<strong>Integration testing</strong>: A level of software testing
where individual units are combined and tested as a group. The purpose
of this level of testing is to expose faults in the interaction between
integrated units.</li>
<li>
<strong>System testing</strong>: A level of the software testing
process where a complete, integrated system is tested. The purpose of
this test is to evaluate whether the system as a whole gives the correct
outputs for given inputs.</li>
</ul>
</div>
</div>
</div>
<p>No matter the type of testing you use, general guidance is to start
by writing any test and make a habit of running tests often.</p>
<ul>
<li>Make improvements where you can, and do your best to include tests
with new code you write even if itâs not feasible to write tests for all
the code thatâs already written.</li>
<li>Make the cases you test as realistic as possible. If for example,
you have dummy data to run tests on you should make sure that data is as
similar as possible to the actual data. If your actual data is messy
with a lot of null values, so should your test dataset be.</li>
</ul>
<p>There are tools available to make writing and running tests easier,
these are known as testing frameworks. Find one you like, learn about
the features it offers, and make use of them.</p>
<p>Writing tests typically encourage researchers to write cleaner, more
modular code as such code is far easier to write tests for, leading to
an improvement in code quality. As well as advantaging individual
researchers testing also benefits research as a whole. It makes research
more reproducible by answering the question âhow do we even know this
code worksâ. To gain an in-depth understanding of different kinds of
tests, please see <a href="https://the-turing-way.netlify.app/reproducible-research/testing.html" class="external-link">Code
Testing</a> chapter in <em>The Turing Way</em>.</p>
</div>
</section><section id="code-review"><h2 class="section-heading">Code Review<a class="anchor" aria-label="anchor" href="#code-review"></a>
</h2>
<hr class="half-width">
<div id="testimonial2" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial2"></a>
</h3>
<div class="callout-content">
<p>The most difficult part of writing code is always to make it
understandable to other people, including yourself a few months down the
track. Thereâs certainly no shame in finding out that your code wasnât
as easy to understand or use as youâd hoped, so donât take it personally
when it happens (which it always does, at least in my experience), but
treat it as an opportunity to improve.</p>
<p><strong>Fernando Perez, <a href="https://fperez.org/py4science/code_reviews.html" class="external-link">Code reviews: the
lab meeting for code</a></strong></p>
</div>
</div>
</div>
<p>A simple objective of the review process is to catch bugs and
elementary errors that might have been missed during the development
phase. Code review can also help improve the overall quality while
ensuring that code is readable and easy to understand. As a group
leader, you can also make sure code is functional and literate as early
as possible, and encourage your students to avoid messy âgood enoughâ
code that causes chaos later.</p>
<p>Code review is often done in pairs, with each reviewer also having
some of their code reviewed by their partner. Doing this can help
programmers to see and discuss issues and alternative approaches to
tasks, and to learn new tips and tricks.</p>
<figure><img src="https://zenodo.org/api/iiif/v2/5c8c70c9-4119-4917-91d1-bc955943f586:328322b3-ab2f-43c8-a4ba-eb6e11d48ac0:reusable-code-garden-with-text.jpg/full/750,/0/default.jpg" alt="Garden of code" class="figure mx-auto d-block"></figure><p><em>The Turing Way project illustration by Scriberia. Used under a
CC-BY 4.0 licence. DOI: 10.5281/zenodo.3332807.</em></p>
<p>There are different methods for code review.</p>
<div class="section level3">
<h3 id="synchronous---pair-programming">Synchronous - Pair Programming<a class="anchor" aria-label="anchor" href="#synchronous---pair-programming"></a>
</h3>
<p><strong>Helping the student go through their scripts, catch errors
and debug side by side</strong></p>
<ul>
<li>The PI sits down with her PhD student who has been writing a
function for cleaning bioinformatics data.</li>
<li>The PI knows Python well and takes the opportunity to discuss code
while helping their student organise the code better.</li>
<li>The student shows the PI some odd errors and so they run some tests
with expected outcomes to find what the problem is and solve it.</li>
<li>The PhD student learns and applies to test practices to help make
code robust.</li>
</ul>
<p>The problem with synchronous coding sessions is making time for it
and whether or not the supervisor has experience with the specific
language.</p>
</div>
<div class="section level3">
<h3 id="synchronous---group-code-tour-or-informal-walkthroughs">Synchronous - Group Code Tour or Informal Walkthroughs<a class="anchor" aria-label="anchor" href="#synchronous---group-code-tour-or-informal-walkthroughs"></a>
</h3>
<p><strong>Narrating code and software steps</strong></p>
<p>The researcher may present their pipeline to describe the logical
steps using documentation, pseudocode, or describing how to run the
code.</p>
<ul>
<li>A postdoc has been working on some analysis that provides statistics
results that he hopes to publish soon. During a lab meeting, the postdoc
presents the steps of the analysis code as logical steps.</li>
<li>The lines of code are shown for those in the meeting that know R,
but the postdoc explains the steps verbally as well for those who donât
understand R.</li>
<li>The group discuss and provides comments on the choices and order of
the analysis pipeline, a PhD student notices a jump in logic that wasnât
picked up previously, and an advanced R user in the lab makes
suggestions about making some parts run faster.</li>
</ul>
<!--
- more detail about the flat hierarchy, not scalable to have the expert to the review
- doesn't need to be an expert
--><p>These sessions do not rely on everyone knowing the language, and it
is the responsibility of the coder to present their work clearly and
logically for everyone to follow. Group discussions can be very
informative for everyone involved and put the analysis under
scrutiny.</p>
<div id="suggestions-for-the-meeting-leader" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="suggestions-for-the-meeting-leader" class="callout-inner">
<h3 class="callout-title">Suggestions for the meeting leader<a class="anchor" aria-label="anchor" href="#suggestions-for-the-meeting-leader"></a>
</h3>
<div class="callout-content">
<ul>
<li>Keep it a safe environment, i.e.Â make sure chastising is relatively
gentle even when deserved (but do point out when code doesnât meet the
required standard â frame it as a learning experience though).</li>
<li>Make sure thereâs a core of vocal participants so it isnât always
you.</li>
<li>Make it clear when youâre presenting yourself that your code isnât
perfect, point out some of those imperfections yourself if the audience
is slow to do so, and do present yourself.</li>
<li>Patiently explain when things are not wrong but just stylistic
differences (but make it clear that some styles are bad, often helpful
e.g.Â asking people to guess what a function returns from its name).</li>
</ul>
<p><em>Shared by Rob Knight with Fernando Perez in the post <a href="https://fperez.org/py4science/code_reviews.html" class="external-link">Code reviews: the
lab meeting for code</a></em></p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="asynchronous---ill-get-back-to-you-on-that">Asynchronous - Iâll get back to you on that<a class="anchor" aria-label="anchor" href="#asynchronous---ill-get-back-to-you-on-that"></a>
</h3>
<p>Making sure everyone is free at the same time for a lab meeting can
be challenging. Hence, asynchronous code review practices are more
suitable for busy supervisors or collaborators in different time
zones.</p>
<p>The asynchronous review process allows others to run the code
themselves using a reproducible environment, or simply reads through the
scripts and share their feedback asynchronously.</p>
<p>Consider a scenario:</p>
<blockquote>
<p>A postdoc has created a model in Python and creates a Binder with all
the dependencies necessary. She sends the file to her supervisor who can
run the code within her browser, no installation is required. The
supervisor can then run the code herself to review it and check the
individual parts over the next week. The supervisor adds a commented
version of the script to the postdocâs repo with a merge request.</p>
</blockquote>
<p>Reviewing code in small chunks incrementally as the project is
developing can help make the code review process a lot more efficient.
Asynchronous feedback removes the time pressure but can be easily
forgotten!</p>
<div id="testimonial3" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial3"></a>
</h3>
<div class="callout-content">
<p>Reviewing more than 400 lines of code (LoC) can have an adverse
impact on your ability to find bugs, and in fact, most are found in the
first 200 lines. - Recommendation from Code Review at Cisco Systems</p>
<p><strong><a href="https://www.atlassian.com/blog/add-ons/code-review-best-practices**" class="external-link">5
code review best practices. Work Life by Atlassian</a>, Usman
Ghani</strong></p>
</div>
</div>
</div>
</div>
</section><section id="multiple-people-can-also-review-the-code-asynchronously."><h2 class="section-heading">Multiple people can also review the code asynchronously.<a class="anchor" aria-label="anchor" href="#multiple-people-can-also-review-the-code-asynchronously."></a>
</h2>
<hr class="half-width">
<div id="callout2" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout2"></a>
</h3>
<div class="callout-content">
<p><a href="https://the-turing-way.netlify.app/reproducible-research/reviewing/reviewing-recommend.html" class="external-link">Turing
Way: Recommendations for Code Reviewing</a></p>
<p>Unlike traditional, âacademic-styleâ peer review, most code review
systems have several advantages: theyâre rarely anonymous, theyâre
public-facing, and without the broker of an editor, contact between
reviewer and reviewee can be direct and rapid. This means code review is
typically a fast, flexible, and interactive process.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Github features to help with code review (Click to see)
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p><strong>Commit changes</strong>: uploading snapshots when the code
changes. The history of all changes are therefore saved and can be
reverted.</p>
<figure><img src="https://i.postimg.cc/MHm8X1zX/Screenshot-2022-02-10-at-16-35-50.png" alt="drawing" width="600" class="figure mx-auto d-block"></figure><p><strong>Branching</strong>: keep a version of the code separate while
making experimental changes or keeping track of collaborative work. Can
try out new functionality or edit in parallel without impacting the code
base.</p>
<figure><img src="https://i.postimg.cc/6p5v0Nb1/Screenshot-2022-02-10-at-18-52-47.png" alt="drawing" width="200" class="figure mx-auto d-block"></figure><p><strong>Pull Request</strong>: Bring the changes made on a branch
over to the main code base. Can be used to request a code review (see
Reviewers on the right panel)</p>
<figure><img src="https://i.postimg.cc/5tgv5Rpm/Screenshot-2022-02-10-at-18-44-07.png" alt="drawing" width="600" class="figure mx-auto d-block"></figure><p><strong>Review</strong>: A pull request can be reviewed and commented
on.</p>
<figure><img src="https://i.postimg.cc/9XDbVyWQ/Screenshot-2022-02-10-at-19-11-58.png" alt="drawing" width="400" class="figure mx-auto d-block"></figure><p><strong>Author: Lydia France (Junior Data Scientist, The Alan Turing
Institute, UK)</strong></p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="reviewing-is-not-about-creating-more-work-nor-the-pi-rewriting-everything">Reviewing is not about creating more work, nor the PI rewriting
everything<a class="anchor" aria-label="anchor" href="#reviewing-is-not-about-creating-more-work-nor-the-pi-rewriting-everything"></a>
</h3>
<p>Instead, it is just another part of peer review and accountability
within the scientific process. It is also an opportunity for everyone to
learn better practices from each other, and solve issues that have
plagued one person for weeks!</p>
<div id="callout3" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout3"></a>
</h3>
<div class="callout-content">
<p><em>Scientists are very aware that their understanding of code
dissipates over time and that this is a large hidden cost. Equally, they
suspect that they spend a lot of time reinventing wheels. They may not
know how code review will help with that, but they hope that it
will.</em></p>
<p><em>One of the mentors expected scientists to overhaul complete code
bases. The advice from one mentor was cogent: if you check the docstring
and write a test every time you touch a method, the code improvements
will accumulate over time with minimal effort.</em></p>
<p><em>Someone who isnât intimately involved with your project should
understand from the module documentation and the comments what you are
trying to do, what approach youâre taking, and why they should expect it
to work.</em></p>
<p><em>Take some time to prepare a presentation about your code that
will answer the above questions even for someone who hasnât read the
code. Youâre more likely to get useful feedback, rather than nitpicking
about syntax, if the audience can see the big picture.</em></p>
<p><em>Keep it a safe environment, i.e.Â make sure chastising is
relatively gentle even when deserved (but do point out when code doesnât
meet the required standard â frame it as a learning experience
though).</em></p>
<p><strong><em>Marian Petre and Greg Wilson. âCode review for and by
scientists: preliminary findings.â (2014).</em></strong></p>
</div>
</div>
</div>
<p>For further considerations in code review, please read <a href="https://the-turing-way.netlify.app/reproducible-research/reviewing.html" class="external-link">Code
Reviewing Process</a> chapter in <em>The Turing Way</em>.</p>
</div>
<div class="section level3">
<h3 id="what-to-look-for-during-code-review">What to look for during Code Review<a class="anchor" aria-label="anchor" href="#what-to-look-for-during-code-review"></a>
</h3>
<p><imzg src="https://the-turing-way.netlify.app/_images/readable-code.jpg" alt="drawing" width="400"></imzg></p>
<p>Reviewing code makes a big difference. Knowledge of the language is
not always necessary!</p>
<p>These are very common, everyone does this.</p>
<p><strong>Bugs/Potential bugs</strong></p>
<ul>
<li>Repetitive code</li>
<li>Code saying one thing, documentation saying another</li>
<li>Off-by-one errors</li>
<li>Making sure each function does one thing only</li>
<li>Lack of tests and sanity checks for what different parts are
doing</li>
<li>Magic numbers (a number hardcoded in the script)</li>
</ul>
<p><strong>Unclear, messy code</strong></p>
<ul>
<li>Bad variable/method names</li>
<li>Inconsistent indentation</li>
<li>The order of the different steps</li>
<li>Too much on one line</li>
<li>Lack of comments and signposting</li>
</ul>
<p><strong>Fragile and non-reusable code</strong></p>
<ul>
<li>Tailor-made and manual steps</li>
<li>Only works with the given data</li>
</ul>
<p><em>Modified from <a href="https://www.cs.swarthmore.edu/~alinen/cs71/labs/lab03.html" class="external-link"><em>What
to look for when code reviewing</em></a></em></p>
</div>
<div class="section level3">
<h3 id="benefits-of-code-review">Benefits of Code Review<a class="anchor" aria-label="anchor" href="#benefits-of-code-review"></a>
</h3>
<div id="testimonial4" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Testimonial<a class="anchor" aria-label="anchor" href="#testimonial4"></a>
</h3>
<div class="callout-content">
<p>In a group of 11 programs developed by the same group of people, the
first 5 were developed without reviews. The remaining 6 were developed
with reviews. After all the programs were released to production, the
first 5 had an average of 4.5 errors per 100 lines of code. The 6 that
had been inspected had an average of only 0.82 errors per 100. Reviews
cut the errors by over 80 percent.</p>
<p><strong><a href="https://www.oreilly.com/library/view/code-complete-2nd/0735619670" class="external-link">Code
Complete</a> by Steve McConnell</strong></p>
</div>
</div>
</div>
<p>The main benefit is finding problems, and finding them early enough
that there arenât frustrating consequences. The penalty for finding a
bug once all the figures have been produced and conclusions drawn, or,
worst-case scenario, after a publication, is much higher than the
penalty for taking the time to review.</p>
<div id="writing-code-collaboratively-also-benefit-directly-your-team-members" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="writing-code-collaboratively-also-benefit-directly-your-team-members" class="callout-inner">
<h3 class="callout-title">Writing code collaboratively also benefit
directly your team members:<a class="anchor" aria-label="anchor" href="#writing-code-collaboratively-also-benefit-directly-your-team-members"></a>
</h3>
<div class="callout-content">
<ul>
<li>Less time redoing work or refactoring</li>
<li>Increased productivity</li>
<li>Greater confidence in own work</li>
<li>Learning better techniques</li>
<li>Reduced time debugging alone</li>
<li>Knowledge exchange and group cohesion</li>
</ul>
</div>
</div>
</div>
<div id="for-a-group-leader-the-benefits-include" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="for-a-group-leader-the-benefits-include" class="callout-inner">
<h3 class="callout-title">For a group leader, the benefits include:<a class="anchor" aria-label="anchor" href="#for-a-group-leader-the-benefits-include"></a>
</h3>
<div class="callout-content">
<ul>
<li>Better understanding of the projects</li>
<li>More maintainable and better-documented code that is easy to
understand and modify</li>
<li>Better insight into any problems with data</li>
<li>Earlier visibility of quality issues</li>
<li>Group reviews reduce the work burden</li>
<li>More robust analysis pipelines that can be reused and modified</li>
<li>High-quality code that can be released</li>
</ul>
</div>
</div>
</div>
<p><strong>Important things to bear in mind:</strong></p>
<p>Code reviews should not be used to evaluate individuals and their
skill levels. An open and safe environment where revealing mistakes and
errors should not come with penalties or shame. Code reviews should also
be done early and often, to normalise this practice in the research
team.</p>
<p>In their book <strong>Peer Reviews in Software: A Practical
Guide</strong>, Karl E. Wiegers says:</p>
<p><em>The temptation to perfect the product before you allow another
pair of eyes to see it. This is an ego-protecting strategy: you wonât
feel embarrassed about your mistakes if no one else sees them. â¦review
[is not] a seal of approval but rather in-process quality-improvement
activity. Such reluctance has several unfortunate consequences. If your
work isnât reviewed until you think itâs complete, you are
psychologically resistant to suggestions for changes.</em></p>
<p>If the program runs, how bad can it be? You are likely to rationalise
away possible bugs because you believe youâve finished and youâre eager
to move on to the next task. Relying on your own desk checking and unit
testing ignores the greater efficiency of a peer review for finding many
defects*</p>
</div>
</section><section id="group-code-writing"><h2 class="section-heading">Group Code Writing<a class="anchor" aria-label="anchor" href="#group-code-writing"></a>
</h2>
<hr class="half-width">
<p>As well as reviewing specific scripts and analyses written by a
single individual, can be very beneficial to solving programming
problems as a team. Setting aside an afternoon to work as a group will
help teach less experienced members of the group and more efficiently
solve very difficult problems.</p>
<p>Groups of people working on a specific problem are often known as
âHackathonsâ in programming. These can last multiple days (hopefully
with downtime!). With very large groups, people can work in pairs or
small groups with delegated parts of the problem to solve and regularly
meet back together to discuss and evaluate. If there is a complex
solution in computational methods that most people in the group need, it
makes sense to find it together.</p>
<p>Similarly, documentation sprints are useful to dedicate time to
regularly bring a codebase to a good minimum standard. Splitting the
task across the team as an event, creating documentation and working
examples for code repos and releasing it can help others use your
computational methods and tools to increase the impact of your work.
Having regularly updated documentation also reduces onboarding time for
new members picking up the shared methods in the lab.</p>
<p>Group work shares the burden and allows knowledge exchange and
support within the team.</p>
<div class="section level3">
<h3 id="continuous-integration">Continuous integration<a class="anchor" aria-label="anchor" href="#continuous-integration"></a>
</h3>
<p>Continuous Integration (CI) is the practice of integrating changes to
a project made by individuals into a main, shared version frequently
(usually multiple times per day). CI is also typically used to identify
any conflicts and bugs that are introduced by changes, so they are found
and fixed early, minimising the effort required to do so. Running tests
regularly also saves humans from needing to do it manually. By making
users aware of bugs as early as possible researchers (if the project is
a research project) do not waste a lot of time doing work that may need
to be thrown away, which may be the case if tests are run infrequently
and results are produced using faulty code. There are many CI service
providers, such as GitHub Actions that come with their own advantages
and disadvantages.</p>
<p><img src="https://the-turing-way.netlify.app/_fig/github-actions.jpg" alt="Continuous Integration with GitHub Actions" class="figure"><em>The Turing Way
project illustration by Scriberia. Used under a CC-BY 4.0 licence. DOI:
10.5281/zenodo.3332807.</em></p>
<p>To learn more about different CI tools and how to use them, please
read the <a href="https://the-turing-way.netlify.app/reproducible-research/ci/ci-options.html" class="external-link">Continuous
Integration</a> chapter in <em>The Turing Way</em>.</p>
</div>
</section><section id="references"><h2 class="section-heading">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<em>The Turing Way</em> Community. (2021). The Turing Way: A
handbook for reproducible, ethical and collaborative research (1.0.1).
Zenodo. <a href="https://doi.org/10.5281/zenodo.5671094" class="external-link">https://doi.org/10.5281/zenodo.5671094</a>.
<a href="https://the-turing-way.netlify.app/reproducible-research/reviewing.html" class="external-link">Code
Reviewing Process</a> Chapter.</li>
<li>Fernando Perez, <a href="https://fperez.org/py4science/code_reviews.html" class="external-link">Code reviews: the
lab meeting for code</a>
</li>
</ul>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ul>
<li>There are many benefits of code review and this should be
implemented and practised in research team culture as early and as
frequently as possible.</li>
<li>Synchronous code review creates opportunities for researchers to get
feedback and learn from others in real-time.</li>
<li>Asynchronous code review is a good practice when working with busy
researchers or collaborators in different time zones.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-27_publications"><p>Content from <a href="27_publications.html">Managing publication</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/27_publications.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why should I make my research objects available?</li>
<li>What open source tools to use for applying data science practices in
bioscience?</li>
<li>How to get your research work cited and invite more contributions to
your project?</li>
<li>How to maintain history of contributions and contributors?</li>
<li>How should I mention and cite the software or data I am using?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the importance of PID.</li>
<li>Know how to make data and code citable</li>
<li>Nudge the use of GitHub/GitLab for open collaboration</li>
<li>Learn why and how to cite the work of others, as well as your own
research output</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><br><br></p>
<section id="what-and-when-to-publish"><h2 class="section-heading">What and when to publish<a class="anchor" aria-label="anchor" href="#what-and-when-to-publish"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/evolution-open-research.png" alt="" class="figure mx-auto d-block"><figcaption>evolution toward open research</figcaption></figure><p>While the output of research projects is usually centred around
publishing a journal article, this format of science communication and
knowledge sharing is increasingly restrictive with the new ways
scientific research is conducted. The requirements from journals
themselves is also expanding, you are now often asked to upload data
sets and code as part of your publication. Releasing data is
increasingly a requirement from funding bodies, and outputs from
research groups can go beyond a single paper, releasing tools and
methods that can be used worldwide.</p>
<p>In general, there are different degrees of openness, depending on
what, when and how research outputs are shared. In order to produce
outputs that can be shared, one should <strong>start preparing for
publication at the beginning of a project</strong>. This overall course
is about getting ready to share code and data in the lab, such that
sharing it with the world is only one click away.</p>
<div class="section level3">
<h3 id="what-can-be-released">What can be released:<a class="anchor" aria-label="anchor" href="#what-can-be-released"></a>
</h3>
<ul>
<li>
<strong>Open Access manuscritps and preprints</strong>: Making the
story and interpretation of your research freely accessible for maximum
use and impact.</li>
<li>
<strong>Open Data</strong>: Documenting and sharing research data
openly for re-use.</li>
<li>
<strong>Open Source Software</strong>: Documenting research code and
routines, and making them freely accessible and available.</li>
<li>
<strong>Open Hardware</strong>: Documenting designs, materials, and
other relevant information related to hardware, and making them freely
accessible and available.</li>
<li>
<strong>Open Notebooks</strong>: An emerging practice, documenting
and sharing the experimental process of trial and error.</li>
</ul>
<blockquote>
<p><a href="https://the-turing-way.netlify.app/reproducible-research/open.html" class="external-link">https://the-turing-way.netlify.app/reproducible-research/open.html</a></p>
</blockquote>
</div>
<div class="section level3">
<h3 id="when-to-release">When to release:<a class="anchor" aria-label="anchor" href="#when-to-release"></a>
</h3>
<ul>
<li>Never: Fully private data and code, unavailable</li>
<li>Pseudo-open â âavailable on requestâ</li>
<li>Released static output parallel to a research paper</li>
<li>Release versioned outputs during the research process in Open online
repository â CRAN, GitHub</li>
<li>Open per default: Collaborative, open science tool with ongoing
development</li>
</ul>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>What are the equivalent of these openness steps for research
manuscript ?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ul>
<li>There were a time when researchers were not sharing their discovery
at all.</li>
<li>The publishing industry made it to a âavailable on requestâ kind of
state.</li>
<li>Open access publication made it a static released.</li>
<li>Preprints are released versions.</li>
<li>blog posts may be considered open per default.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="how-can-it-be-released">How can it be released<a class="anchor" aria-label="anchor" href="#how-can-it-be-released"></a>
</h3>
<p>This relates to the license used: what will people be allowed to do
with your released research outputs.</p>
<p>Without going into details, different types of outputs should be
released using different licenses. Licenses are usually of different
forms, giving the rights to use without restriction (public domain),
allowing to use as long as the original authors are credited (with
attribution). In addition one can restrict re-use the work as long as
the new work is released with a similar license (weak copyleft) or the
same license (hard copyleft).</p>
<p>Some license also restrict the use to non-commercial use or
preventing derivatives.</p>
<blockquote>
<p>See <a href="https://the-turing-way.netlify.app/reproducible-research/licensing" class="external-link">https://the-turing-way.netlify.app/reproducible-research/licensing</a>
and linked tools for details.</p>
</blockquote>
<p>In addition, one should think about where things are released (how
long will it be available, how will it be found), and how to indicate
how people should credit your work.</p>
<div id="open-or-private" class="callout testimonial">
<div class="callout-square">
<i class="callout-icon" data-feather="heart"></i>
</div>
<div id="open-or-private" class="callout-inner">
<h3 class="callout-title">Open or Private?<a class="anchor" aria-label="anchor" href="#open-or-private"></a>
</h3>
<div class="callout-content">
<p>Researchers often worry that they need to hide their code to prevent
others stealing it.</p>
<blockquote>
<p>âAfter giving talks about open science Iâve sometimes been approached
by skeptics who say, âWhy would I help out my competitors by sharing
ideas and data on these new websites? Isnât that just inviting other
people to steal my data, or to scoop me? Only someone naive could think
this will ever be widespread.â As things currently stand, thereâs a lot
of truth to this point of view. But itâs also important to understand
its limits. What these skeptics forget is that they already freely share
their ideas and discoveries, whenever they publish papers describing
their own scientific work. Theyâre so stuck inside the
citation-measurement-reward system for papers that they view it as a
natural law, and forget that itâs socially constructed. Itâs an
agreement. And because itâs a social agreement, that agreement can be
changed. All thatâs needed for open science to succeed is for the
sharing of scientific knowledge in new media to carry the same kind of
cachet that papers do todayâ</p>
<p>Nielsen, M. Reinventing Discovery: The New Era of Networked Science.
Princeton University Press, 2011.</p>
</blockquote>
<p><a href="https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3000246" class="external-link">https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3000246</a></p>
</div>
</div>
</div>
</div>
</section><section id="making-your-research-object-citable"><h2 class="section-heading">Making your research object citable<a class="anchor" aria-label="anchor" href="#making-your-research-object-citable"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="getting-a-doi">Getting a doi<a class="anchor" aria-label="anchor" href="#getting-a-doi"></a>
</h3>
<figure><img src="fig/DOI.jpg" alt="This image shows three boxes with materials on top. The main box in the middle has 'identifiers' written on it with three discs on top of it that are labelled 'data sets'. Both boxes by their side have journal articles on top of them. An arrow on the top of the image points to these images as being 'Digital Object Identifiers'. There is text at the bottom of the image which says 'Persistent', 'Unique', 'Trusted'." width="500" class="figure mx-auto d-block"></figure><p><em>Digital Object Identifiers or DOIs are persistent, unique and
trusted. <em>The Turing Way</em> project illustration by Scriberia. Used
under a CC-BY 4.0 licence. DOI: <a href="https://doi.org/10.5281/zenodo.3332807" class="external-link">10.5281/zenodo.3332807</a>.</em></p>
<p>DOIs are alphanumerical unique and persistent identifiers with a
permanent web address for different research objects that can be cited
by you and other researchers. Each pre-print and publication is
published with a DOI, but independent of the paper, different research
objects can be published online on servers that offer DOIs at any stage
of your research.</p>
<p>Repositories usually collect the citation information before
providing a DOI, so that the metadata associated with the DOI is
indexed. Necessary information includes a title and a list of
authors.</p>
<p>You may want to read the <a href="https://the-turing-way.netlify.app/communication/citable.html" class="external-link">Making
Research Objects Citable</a> chapter in <em>The Turing Way</em> Guide to
Communication.</p>
</div>
<div class="section level3">
<h3 id="authors-and-contributors">Authors and contributors<a class="anchor" aria-label="anchor" href="#authors-and-contributors"></a>
</h3>
<p>In order to get correct metadata, give credits to all contributors
and actually have the rights to make things public, you need to collect
information and agreements from all authors. As usual, it is best to get
it as soon as possible in the research process; and as usual you may
well have postponed it until the latest moment.</p>
<p>It is advisable to use a spreadsheet to collect information (names,
affiliation, orcid number). The Tenzing spreadsheet can be used, as it
will allow the program to create author list automatically from the
spreadsheet, as well as collect information on each author
contribution.</p>
<div id="callout1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>It might be particularly important to credit all people who have
indeed participated in the research. The use of a contribution statement
in the paper (see <a href="https://tenzing.club" class="external-link">https://tenzing.club</a>) or contribution
section in the repository readme can be useful (see <a href="https://allcontributors.org" class="external-link">https://allcontributors.org</a>).</p>
<p>For contributors who will not be in the author list of the main
paper, one can make them authors of the published dataset/code, and/or
add them in the aknowledgements. The first solution has the advantage to
be sent to orcid (if people have an orcid account) such that the
contribution is more visible and measurable.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="citation-file-format">Citation File Format<a class="anchor" aria-label="anchor" href="#citation-file-format"></a>
</h3>
<p>The Citation File Format provides citation metadata, for software or
datasets, in plaintext files that are easy to read by both humans and
machines.</p>
<p>Adding a <code>CITATION.cff</code> file to your folder means it can
be cited when others use it, increasing recognition for your work and
your research projectâs impact.</p>
<p>See more at <a href="https://the-turing-way.netlify.app/communication/citable/citable-cff.html" class="external-link">The
Turing Way: CITATION.cff</a></p>
<p><a href="https://the-turing-way.netlify.app/_images/software-credit.jpg" class="external-link">https://the-turing-way.netlify.app/_images/software-credit.jpg</a></p>
</div>
<div class="section level3">
<h3 id="citable-data">Citable data<a class="anchor" aria-label="anchor" href="#citable-data"></a>
</h3>
<p>Data is best published in domain specific repositories. Look for a
data repository at <a href="https://www.re3data.org" class="external-link uri">https://www.re3data.org</a>.</p>
<p>Costs associated with data publication can be important, so it is
best to plan ahead, so you are sure you have the resources needed.
Whether the raw or some derived data should be published openly will
depend on costs and may depend on legislations.</p>
<p>See <a href="https://the-turing-way.netlify.app/communication/citable.html" class="external-link">https://the-turing-way.netlify.app/communication/citable.html</a>
for details.</p>

</div>
<div class="section level3">
<h3 id="citable-code">Citable Code<a class="anchor" aria-label="anchor" href="#citable-code"></a>
</h3>
<p>For computational projects, releasing your work in an open repository
has parallels with publications.
<img src="fig/parallel.png" alt="similarity between publication and repository use" width="800" class="figure"></p>
<p>While it is possible to cite software from GitHub (see below using a
Citation File Format file), it is better to cite an archived version of
the software (a code release). You can release code and data associated
with a research article as a series of files/folders. You could for
example bundle folders into a <code>.zip</code> file and upload it to
Zenodo.</p>
<p>The possibility to get integration with zenodo from GitHub or GitLab,
makes it easy to get DOIs for each version of the software. These
automated archive of release may even read the CITATION.cff file to
create metadata. This is explained in more detail in the Turing way book
at <a href="https://the-turing-way.netlify.app/communication/citable/citable-cite.tml" class="external-link uri">https://the-turing-way.netlify.app/communication/citable/citable-cite.tml</a></p>

<!--
## Code release

- Packaging (R/Python package)
- Environment (conda, binder, docker)
- Code release (Zenodo, Figshare)
  - Connecting to Zenodo
  - Citable code: CFF

## Data Release 
- Institute's policy for archival
- Public data deposit like Data Dryad)
- Contributing to open tools

## Manuscript, Pre-prints and Peer Review

- Best practices and examples

## Supporting Materials

- GitHub repo
- Interactive page for your project
- Jupyter/Binder

## Conclusion
- What gaps have we filled in this section
- Project management overview 

## Resources for taking this to next level

- Turing Data Stories, The Turing Way, Shiny app

## Next Steps

In addition to the referenced linked under different sections in this lesson, please see the following references:
- [Add recommendations from 1:1 interviews, open communities such as Open Life Science and The Turing Way.]
-->
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Add a CITATION.cff file to your Github repository.</li>
<li>Create a tenzing spreadsheet to record contribution.</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Online Persistent Identifiers or Digital Object Identifiers are
useful for releasing and citing different versions of research
objects.</li>
<li>Planning for publication will change your workflow, making these
decisions early and allocating time to implement your decision is
key</li>
</ul>
</div>
</div>
</div>
<p><br><br><br><br></p>
<hr>
<p><br><br></p>
</div>
</section><section id="citing-others"><h2 class="section-heading">Citing others<a class="anchor" aria-label="anchor" href="#citing-others"></a>
</h2>
<hr class="half-width">
<p>You should cite research objects directly in the paper in places
where it is relevant. This is a commonly practiced way of citing
publications and is valid for citing other research components like
data, hardware and software.</p>
<p>A citation includes the following information: - Author - Title -
Year of publication - Publisher (for data, this is often the data
repository where it is housed) - Version (if indicated) - Access
information (a URL or DOI)</p>
<p>You should use a citation manager to collect and use this
information, <a href="https://www.zotero.org" class="external-link">zotero</a> is the free and
open source software option.</p>
<p>If the author of the research object did not provide a easy way to
collect the information, try to contact them. You can refer them to the
[turing way chapter] (<a href="https://the-turing-way.netlify.app/communication/citable.html" class="external-link uri">https://the-turing-way.netlify.app/communication/citable.html</a>)
, which is also a source for this course.</p>
</section><section id="citing-yourself"><h2 class="section-heading">Citing yourself<a class="anchor" aria-label="anchor" href="#citing-yourself"></a>
</h2>
<hr class="half-width">
<p>Make your research objects are citable, as described later. Then do
not forget to insert citation of these research objects in your own
papers. Provide data, software and hardware availability statement, for
example:</p>
<p>The data/software/harware that support the findings will be available
in [repository name] at [URL / DOI] following a [6 month] embargo from
the date of publication, to allow for the commercialization of research
findings.â</p>
<p>Note that you can often reserve a DOI without publishing your data
and code, if you want for the paper to come out first.</p>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ul>
<li>Cite code and data</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-28_conclusion"><p>Content from <a href="28_conclusion.html">Conclusion and feedback</a></p>
<hr>
<p> Last updated on 2023-09-12 | 
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/episodes/28_conclusion.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the most important take home messages?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Overview of what has been learned</li>
<li>consolidation of learning</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>Feedback, questions</p></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/managing-computational-projects/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/managing-computational-projects/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/managing-computational-projects/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/managing-computational-projects/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:msharan@turing.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.13.3" class="external-link">sandpaper (0.13.3)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.6.1" class="external-link">pegboard (0.6.1)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.3.0" class="external-link">varnish (0.3.0)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/managing-computational-projects/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/managing-computational-projects/aio.html",
  "identifier": "https://carpentries-incubator.github.io/managing-computational-projects/aio.html",
  "dateCreated": "2021-08-20",
  "dateModified": "2023-09-26",
  "datePublished": "2023-09-26"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

